---
title: "Ago1_known_novel_miRNA_target_enrichment_in_peaks"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ChIPseeker)
library(dplyr)
library(magrittr)
library(BSgenome)
library(data.table)
library(GenomicFeatures)
library(GenomicRanges)
library(ggplot2)
library(soGGi)
```


```{r}
##seed search fun 
fetchSequencesForClIP <- function(peaks,reSize=NULL,fasta,add5=0,add3=0){
  swd <- Rsamtools::FaFile(fasta)
  
  fastaLens <- swd %>%  seqlengths
  start(peaks) <- start(peaks)-add5
  end(peaks) <- end(peaks)+add3
  sees <- seqlevels(peaks)
  seqlengths(peaks) <- fastaLens[match(sees,names(fastaLens),incomparables = 0)]
  
  
  
  Boundaries <- GRanges(seqlevels(peaks),IRanges(1,seqlengths(peaks)))
  
  if(!is.null(reSize)){
    resizePeaks <- resize(peaks, reSize, fix="center", use.names=TRUE)
    
  }else{
    resizePeaks <- peaks
  }
  names(resizePeaks) <- paste0(seqnames(resizePeaks),":",start(resizePeaks),"-",end(resizePeaks))
  #temp <- subsetByOverlaps(resizePeaks,Boundaries,type=c("within"))
  resizePeaks <- trim(resizePeaks)##added this line cause off error when extending going ut of bounds
  temp <- findOverlaps(resizePeaks,Boundaries,type=c("within"))
  temp2 <- Rsamtools::getSeq(Rsamtools:::FaFile(fasta),resizePeaks[temp@from])
  names(temp2) <- names(resizePeaks[temp@from])
  temp2
}



annotatePeaksWithPatterns  <- function(peaks,fasta,patterns,resize=64,add5=0,add3=0,verbose=FALSE,checkReverse=TRUE){
  require(GenomicRanges)
  require(magrittr)
  require(Biostrings)
  require(Rsamtools)
  
  if(verbose) message("Reading in peaks....",appendLF = FALSE)
  if(is(peaks,"GRanges")){
    peaks <- peaks
    peaks$originalPeak <-  paste0(seqnames(peaks),":",start(peaks),"_",end(peaks), "_", strand(peaks))
  }else if(is(peaks,"character")){
    if(!file.exists(peaks))stop(paste0("The file ",peaks," does not exist"))
    peaks <- read.table(peaks, header = TRUE)
    peaks <- makeGRangesFromDataFrame(peaks,
                                      keep.extra.columns=TRUE,
                                      ignore.strand=FALSE,
                                      seqinfo=NULL,
                                      seqnames.field="seqnames",
                                      start.field="start",
                                      end.field="end",
                                      strand.field="strand",
                                      starts.in.df.are.0based=FALSE)
    peaks$originalPeak <-  paste0(seqnames(peaks),":",start(peaks),"_",end(peaks), ":", strand(peaks))
    
  }
  if(verbose) message("...done")
  if(verbose) message("Read in ",length(peaks)," peaks")
  
  
  if(!file.exists(fasta))stop(paste0("The file ",fasta," does not exist"))
  if(verbose) message("Indexing FASTA...",appendLF = FALSE)
  indexFa(fasta)
  if(verbose) message("...done")
  
  
  if(verbose) message("Aligning seqlevels across peaks and FASTA...",appendLF = FALSE)
  fastaLens <- FaFile(fasta) %>% seqlengths
  sees <- seqlevels(peaks)
  seqlengths(peaks) <- fastaLens[match(sees,names(fastaLens),incomparables = 0)]
  if(verbose) message("...done")
  
  if(verbose) message("Removing invalid peaks which are outside contig boundaries...",appendLF = FALSE)
  Boundaries <- GRanges(seqlevels(peaks),IRanges(1,seqlengths(peaks)))
  resizePeaks <- resize(peaks, resize, fix="center", use.names=TRUE)
  resizePeaks <- trim(resizePeaks)
  names(resizePeaks) <- paste0(seqnames(peaks),":",start(peaks),"_",end(peaks), "_", strand(peaks)) #rename accoridng to original (not extended) peakID to be able
  temp <- findOverlaps(resizePeaks,Boundaries,type=c("within"))
  validPeaks <- resizePeaks[temp@from]
  # validPeaks$originalPeak <- names(validPeaks)
  
  
  ##
  # Here we make a name for extended peaks...so we know its extended coordinates even if we change this GRanges!
  ##
  validPeaks$extendedPeak <- paste0(seqnames(validPeaks),":",start(validPeaks),"_",end(validPeaks), "_", strand(validPeaks))
  if(verbose) message("...done")
  if(verbose) message("Removed in ",length(validPeaks)-length(resizePeaks)," peaks")
  
  if(verbose) message("Retrieving sequence from under peaks....",appendLF = FALSE)
  myRes <- getSeq(Rsamtools:::FaFile(fasta),validPeaks)
  
  ##
  ## We set myRes (the reverse complement) names by extended peak names.. 
  ##
  names(myRes) <- validPeaks$extendedPeak
  
  validPeaks$seqUnderExtendedPeak <- myRes
  if(verbose) message("...done")
  
  if(verbose) message("Retrieving patterns to search for....",appendLF = FALSE)
  if(length(patterns) > 1){
    pattern=patterns
  }else{
    motifSeq <- readDNAStringSet(patterns)
    pattern <- as.character(motifSeq)
  }
  if(verbose) message("...done")
  if(verbose) message("Read in ",length(pattern)," patterns")
  
  
  # rm(fixedInRegion)
  if(verbose) message("Searching for ",length(pattern)," patterns in peaks....")
  mergePeaksAndSites <- patternCallList <- list()
  for(i in 1:length(pattern)){
    if(verbose) message("Search for ",pattern[i])
    peaks_Sites <- validPeaks
    fixedInRegion <- vmatchPattern(pattern=pattern[i],subject=myRes) %>% unlist()
    
    
    startOfPmatch <- resize(fixedInRegion,width=1,fix="start") %>% unname %>% as.character
    
    ##
    ## Pos is made from names from original sequence we scanned in..which came from extened peak coordnates. 
    ##  or..Names contains extended position and strand information because myRes did in its name.
    ##
    pos <- matrix(unlist(strsplit(gsub(".*:","",names(fixedInRegion)),"_")),ncol=3,byrow =T)
    seq <- gsub(":.*","",names(fixedInRegion))
    
    ###
    ## Here we lose strand information from our GRanges!! 
    ## sitesGRpos are from Pos strand based on name... but they dont have strand information in them !! 
    ## That why we need to reverse complement.
    ## We could add back strand information somewhere here but doesnt matter as we take care of it ourselves
    
    ## I checked a few sites in IGV from original FASTA and all good! And now we can remember why too :)
    ###
    sitesGRpos <- GRanges(seq[pos[,3] =="+"],
                          IRanges(as.numeric(pos[pos[,3] =="+",1])+start(fixedInRegion[pos[,3] =="+"])-1,
                                  as.numeric(pos[pos[,3] =="+",1])+end(fixedInRegion[pos[,3] =="+"])-1),startOfPmatch=startOfPmatch[pos[,3] =="+"])
    sitesGRpos$PeakID <- names(fixedInRegion[pos[,3] =="+"])
    
    sitesGRneg <- GRanges(seq[pos[,3] =="-"],
                          IRanges(as.numeric(pos[pos[,3] =="-",2])-end(fixedInRegion[pos[,3] =="-"])+1,
                                  as.numeric(pos[pos[,3] =="-",2])-start(fixedInRegion[pos[,3] =="-"])+1),startOfPmatch=startOfPmatch[pos[,3] =="-"])
    sitesGRneg$PeakID <- names(fixedInRegion[pos[,3] =="-"])
    
    ## So we need to reverse complement based on
      sitesFA <- c(fetchSequencesForClIP(sitesGRpos,reSize=NULL,fasta),reverseComplement(fetchSequencesForClIP(sitesGRneg,reSize=NULL,fasta)))
      sitesFAExtended <- c(fetchSequencesForClIP(sitesGRpos,reSize=NULL,fasta,add5=add5,add3=add3),
                           reverseComplement(fetchSequencesForClIP(sitesGRneg,reSize=NULL,fasta,add5=add5,add3=add3))
      )
      sitesGR <- suppressWarnings(c(sitesGRpos,sitesGRneg))

    # getSeq(Rsamtools:::FaFile(fasta),sitesGR)
    sitesGR$Seq <- sitesFA
    sitesGR$SeqExtended <- sitesFAExtended
    sitesGR$siteOfPattern <- granges(sitesGR) %>% unname %>% as.character
    sitesGR$centerOfPattern <-  resize(sitesGR,width=1,fix="center") %>% unname %>% as.character
    dfSitesGR <- as.data.frame(sitesGR)
    # peaks_Sites$extendedRanges <- resize(granges(peaks_Sites), reSize, fix="center", use.names=TRUE)
    # peaks_Sites$peakNames <- paste0(seqnames(peaks_Sites$extendedRanges),":",start(peaks_Sites$extendedRanges),"-",end(peaks_Sites$extendedRanges))
    peaks_SitesDF <- as.data.frame(peaks_Sites)
    
    mergePeaksAndSites[[i]] <- merge(peaks_SitesDF,dfSitesGR,by.x="extendedPeak",by.y="PeakID",all=FALSE)
    patternCallList[[i]] <- mergePeaksAndSites[[i]] %>% dplyr::select(Seq,SeqExtended,startOfPmatch,centerOfPattern,siteOfPattern,originalPeak,extendedPeak)
    rm(peaks_SitesDF)
    rm(peaks_Sites)
  }
  #if(verbose) message("....finished search")
  names(mergePeaksAndSites) <- names(patternCallList) <- as.character(pattern)
  return(list(mergePeaksAndSites,patternCallList))
}
```

##Get peak ranges and control ranges

```{r}
#read in all peaks, with filtering stats/etc added
all <- read.delim("/Users/kathryn/Reprocess_all_paper_datasets/Supp_tables/TableS3_final.txt", sep = "\t", header = TRUE)

#make GR for seed search
allGR <- makeGRangesFromDataFrame(all,
                                     keep.extra.columns=TRUE,
                                     ignore.strand=FALSE,
                                     seqinfo=NULL,
                                     seqnames.field="seqnames",
                                     start.field="start",
                                     end.field="end",
                                     strand.field="strand",
                                     starts.in.df.are.0based=FALSE)

allctrlGR <- GenomicRanges::shift(allGR, 4000) 

gtffile <- "/Users/kathryn/Aedes-aegypti-LVP_AGWG_BASEFEATURES_AaegL5.2.gtf"

TxDb <- makeTxDbFromGFF(gtffile)


allctrlanno <- annotatePeak(allctrlGR ,
                             TxDb=TxDb,
                            sameStrand=FALSE,
                            genomicAnnotationPriority = c("3UTR", "5UTR", "Exon",
                                                           "Intron", "Promoter", "Downstream", "Intergenic"),
                          overlap = "all")

allctrlannodf <- as.data.frame(allctrlanno)

no_ann <- allctrlGR[!allctrlGR$peakID %in% allctrlannodf$peakID,]
no_ann <- as.data.frame(no_ann)
allctrldf <- rbind(no_ann, allctrlannodf)

allctrlGR <- makeGRangesFromDataFrame(allctrldf,
                                    keep.extra.columns=TRUE,
                                     ignore.strand=FALSE,
                                     seqinfo=NULL,
                                     seqnames.field="seqnames",
                                     start.field="start",
                                     end.field="end",
                                     strand.field="strand",
                                     starts.in.df.are.0based=FALSE)
```

##Search top 10 most abundant known miRNAs in mosquitoes and cells 

```{r}
res <- annotatePeaksWithPatterns(peaks=allGR,
                                  fasta="/Users/kathryn/mirdeep2_master/Aedes-aegypti-LVP_AGWG_CHROMOSOMES_AaegL5.fa",
                                patterns="/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/top10fam_known_miRNA_6mer_target.fa",
                                  resize=406,
                                  add5=1,
                                  add3=1,
                                  verbose=TRUE)

res <- res[[1]]
res <- rbindlist(res)

#make sure are only counting pattern matches 1X
#if seqeunce searched AND pattern position are eaxaclty the same, is a duplicate;

res$search_seq_pattern <- paste0(res$seqUnderExtendedPeak, "_", res$siteOfPattern)
nrow(res)
length(unique(res$search_seq_pattern))
# 108465, good, all are unqiue
res$search_seq_pattern <- NULL

#but, could still have dups if extended seq of one peak overlaps with the extended peak of anotheer sequence and then a match is in both; 
#if site of pattern and ex seq is the same, show only the entry with the start of pMatch closest to 0 n
res$group <- paste0(res$SeqExtended, "_", res$siteOfPattern)
length(unique(res$group)) #66228
res$startOfPmatch <- as.numeric(res$startOfPmatch)
res$filt <- duplicated(res$group)
res_Ago1_known <- subset(res, res$filt == FALSE)
nrow(res_Ago1_known)


res_Ago1_known_ctrl <- annotatePeaksWithPatterns(peaks=allctrlGR,
                              fasta="/Users/kathryn/mirdeep2_master/Aedes-aegypti-LVP_AGWG_CHROMOSOMES_AaegL5.fa",
patterns="/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/top10fam_known_miRNA_6mer_target.fa",
                                  resize=406,
                                  add5=1,
                                  add3=1,
                                  verbose=TRUE)

res_Ago1_known_ctrl <- res_Ago1_known_ctrl[[1]]
res_Ago1_known_ctrl <- rbindlist(res_Ago1_known_ctrl)

#make sure are only counting pattern matches 1X
#if seqeunce searched AND pattern position are eaxaclty the same, is a duplicate;

res_Ago1_known_ctrl$search_seq_pattern <- paste0(res_Ago1_known_ctrl$seqUnderExtendedPeak, "_", res_Ago1_known_ctrl$siteOfPattern)
nrow(res_Ago1_known_ctrl)
length(unique(res_Ago1_known_ctrl$search_seq_pattern))
# 101755, good, all are unqiue
res_Ago1_known_ctrl$search_seq_pattern <- NULL

res_Ago1_known_ctrl$group <- paste0(res_Ago1_known_ctrl$SeqExtended, "_", res_Ago1_known_ctrl$siteOfPattern)
length(unique(res_Ago1_known_ctrl$group)) #62644
res_Ago1_known_ctrl$startOfPmatch <- as.numeric(res_Ago1_known_ctrl$startOfPmatch)
res_Ago1_known_ctrl$filt <- duplicated(res_Ago1_known_ctrl$group)
res_Ago1_known_ctrl <- subset(res_Ago1_known_ctrl, res_Ago1_known_ctrl$filt == FALSE)
nrow(res_Ago1_known_ctrl)


Ago1_aegypti_seed_known_ctrl <- subset(res_Ago1_known_ctrl$startOfPmatch, res_Ago1_known_ctrl$aegypti_Ago1_BCsample > 0 & res_Ago1_known_ctrl$aegypti_Ago2_BCsample == 0)
Ago1_aegypti_seed_known <- subset(res_Ago1_known$startOfPmatch, res_Ago1_known$aegypti_Ago1_BCsample > 0 & res_Ago1_known$aegypti_Ago2_BCsample ==0 )

Ago1_Aag2_seed_known_ctrl <- subset(res_Ago1_known_ctrl$startOfPmatch, res_Ago1_known_ctrl$Aag2_Ago1_BCsample > 0 & res_Ago1_known_ctrl$Aag2_Ago2_BCsample ==0)
Ago1_Aag2_seed_known <- subset(res_Ago1_known$startOfPmatch, res_Ago1_known$Aag2_Ago1_BCsample > 0 & res_Ago1_known$Aag2_Ago2_BCsample ==0)


#get GRs with the same peak subset for calculating specific coverage
Ago1_aegypti_ranges <- subset(res_Ago1_known, res_Ago1_known$aegypti_Ago1_BCsample > 0 & res_Ago1_known$aegypti_Ago2_BCsample ==0 )
Ago1_aegypti_ctrl_ranges <- subset(res_Ago1_known_ctrl, res_Ago1_known_ctrl$aegypti_Ago1_BCsample > 0 & res_Ago1_known_ctrl$aegypti_Ago2_BCsample == 0)


Ago1_Aag2_ranges <- subset(res_Ago1_known, res_Ago1_known$Aag2_Ago1_BCsample > 0 & res_Ago1_known$Aag2_Ago2_BCsample ==0 )
Ago1_Aag2_ctrl_ranges <- subset(res_Ago1_known_ctrl, res_Ago1_known_ctrl$Aag2_Ago1_BCsample > 0 & res_Ago1_known_ctrl$Aag2_Ago2_BCsample ==0 )


####aegypti Ago1 known
Ago1_aegyptik_GR <-  GRanges(seqnames = Ago1_aegypti_ranges$seqnames.x, strand = Ago1_aegypti_ranges$strand.x,
                          ranges = IRanges(start =Ago1_aegypti_ranges$start.x, width =Ago1_aegypti_ranges$width.x ))

ToPlot_aegypti_Ago1kalt <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_aegyptiAgo1.bw",Ago1_aegyptik_GR  ,format = "bigwig")
f <- plotRegion(ToPlot_aegypti_Ago1kalt,outliers = 0.05,groupBy = "Sample")
aegypti_Ago1kalt <- f$data

###search ctrl ranges seperately
Ago1_aegyptik_ctrlGR <-  GRanges(seqnames = Ago1_aegypti_ctrl_ranges$seqnames.x, strand = Ago1_aegypti_ctrl_ranges$strand.x,
                             ranges = IRanges(start =Ago1_aegypti_ctrl_ranges$start.x, width =Ago1_aegypti_ctrl_ranges$width.x ))

ToPlot_ctrl_aegypti_kAgo1 <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_aegyptiAgo1.bw",Ago1_aegyptik_ctrlGR,format = "bigwig")
ctrl_aegypti_kAgo1_peaks  <- plotRegion(ToPlot_ctrl_aegypti_kAgo1,outliers = 0.05)
ctrl_aegypti_kAgo1_peaks <- ctrl_aegypti_kAgo1_peaks$data
ctrl_aegypti_kAgo1_peaks$Sample <- c("ctrl_peaks") 

aegypti_Ago1kalt  <- rbind(ctrl_aegypti_kAgo1_peaks, aegypti_Ago1kalt )


######
Ago1_Aag2k_GR <-  GRanges(seqnames = Ago1_Aag2_ranges$seqnames.x, strand = Ago1_Aag2_ranges$strand.x,
                          ranges = IRanges(start = Ago1_Aag2_ranges$start.x, width =Ago1_Aag2_ranges$width.x ))

ToPlot_Aag2_Ago1kalt <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_Aag2Ago1.bw",Ago1_Aag2k_GR ,format = "bigwig")
f <- plotRegion(ToPlot_Aag2_Ago1kalt,outliers = 0.05,groupBy = "Sample")
Aag2_Ago1kalt <- f$data

#ctrl ranges 
Ago1_Aag2k_ctrlGR <-  GRanges(seqnames = Ago1_Aag2_ctrl_ranges$seqnames.x, strand = Ago1_Aag2_ctrl_ranges$strand.x,
                              ranges = IRanges(start = Ago1_Aag2_ctrl_ranges$start.x, width =Ago1_Aag2_ctrl_ranges$width.x ))

ToPlot_ctrl_Aag2_Ago1kalt <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_Aag2Ago1.bw",Ago1_Aag2k_ctrlGR ,format = "bigwig")
ctrlAag2_Ago1kalt_peaks <- plotRegion(ToPlot_ctrl_Aag2_Ago1kalt ,outliers = 0.05)
ctrlAag2_Ago1kalt_peaks <- ctrlAag2_Ago1kalt_peaks$data
ctrlAag2_Ago1kalt_peaks$Sample <- c("ctrl_peaks")

Aag2_Ago1kalt <- rbind(ctrlAag2_Ago1kalt_peaks, Aag2_Ago1kalt)

##now combine cov with pattern matches and graph
nf <- c("Aag2_Ago1kalt", "aegypti_Ago1kalt")
norm_f <- do.call("list",mget(nf))
cov <- lapply(norm_f, function(x) subset(x, x$xIndex >= 1400 & x$xIndex <=1600))
covdf <- rbindlist(cov, idcol = TRUE)
covdf$xIndex <- covdf$xIndex - 1500
col.names <- colnames(covdf)
col.names <- gsub(".id", "ranges", col.names)
col.names <- gsub("xIndex", "position", col.names)
covdf <- setNames(covdf, col.names)
cells_ago1 <- subset(covdf, covdf$ranges=="Aag2_Ago1kalt")
cells_ago1 <- subset(cells_ago1, cells_ago1$Sample=="ctrl_peaks" | cells_ago1$Sample=="cat_Aag2Ago1.bw")
aegypti_ago1 <- subset(covdf, covdf$ranges=="aegypti_Ago1kalt")
aegypti_ago1 <- subset(aegypti_ago1, aegypti_ago1$Sample=="ctrl_peaks" | aegypti_ago1$Sample=="cat_aegyptiAgo1.bw")

aegypti_ago1$merge <- ifelse(aegypti_ago1$Sample=="cat_aegyptiAgo1.bw" , paste0(aegypti_ago1$position, "Ago1_aegypti_seed_known"), paste0(aegypti_ago1$position, "Ago1_aegypti_seed_known_ctrl"))

cells_ago1$merge <- ifelse(cells_ago1$Sample=="cat_Aag2Ago1.bw" , paste0(cells_ago1$position, "Ago1_Aag2_seed_known"), paste0(cells_ago1$position, "Ago1_Aag2_seed_known_ctrl"))
#probably want to norm counts and then get overall freq 


#make list of pattern seed matches, (as above) and then count all together for graphing 
Ago1known <- grep("_seed_known", names(.GlobalEnv),value=TRUE)
Ago1known <- grep("Ago1_", Ago1known, value=TRUE)
Ago1known <- do.call("list",mget(Ago1known))
Ago1known <- lapply(Ago1known, as.data.frame)
Ago1known <- lapply(Ago1known, setNames, nm="position")
Ago1known_counts <- lapply(Ago1known, function(x) x %>% group_by(position) %>% summarize(count=n()))
Ago1known_counts<- lapply(Ago1known_counts, function(x) subset(x, x$position <= 300 & x$position >=100))

#make master df by sample name
Ago1df <- rbindlist(Ago1known_counts, idcol = TRUE)
Ago1df$position <- Ago1df$position - 200
Aag2Ago1k <- subset(Ago1df, Ago1df$.id=="Ago1_Aag2_seed_known_ctrl" | Ago1df$.id=="Ago1_Aag2_seed_known")
Aag2Ago1k$merge <- paste0(Aag2Ago1k$position, Aag2Ago1k$.id)

aegyptiAgo1k <- subset(Ago1df, Ago1df$.id=="Ago1_aegypti_seed_known_ctrl" | Ago1df$.id=="Ago1_aegypti_seed_known")
aegyptiAgo1k$merge <- paste0(aegyptiAgo1k$position, aegyptiAgo1k$.id)

#merge 
aegypti_ago1 <- merge(aegyptiAgo1k, aegypti_ago1, by="merge")
cells_ago1 <- merge(Aag2Ago1k, cells_ago1, by="merge")


#
aegypti_ago1 <- aegypti_ago1 %>%
  group_by(Sample) %>%
  mutate( counts_norm = (count*Score))

aegypti_ago1 <- aegypti_ago1 %>%
  group_by(Sample) %>%
  mutate(freq_norm = counts_norm/(sum(counts_norm)))

cells_ago1 <- cells_ago1 %>%
  group_by(Sample) %>%
  mutate( counts_norm = (count*Score))

cells_ago1 <-cells_ago1 %>%
  group_by(Sample) %>%
  mutate(freq_norm = counts_norm/(sum(counts_norm)))

aegypti_ago1k  <-split(aegypti_ago1, aegypti_ago1$.id)
aegypti_ago1k  <- lapply(aegypti_ago1k, function(x) x[order(x$position.x),])
aegypti_ago1k_spline <-lapply(aegypti_ago1k, function(x) spline(x$freq_norm))
aegypti_ago1k_spline <- rbindlist(aegypti_ago1k_spline, idcol = TRUE)
aegypti_ago1k_spline$x <- aegypti_ago1k_spline$x - 101
col.names <- colnames(aegypti_ago1k_spline)
col.names <- gsub(".id", "sample", col.names)
aegypti_ago1k_spline <- setNames(aegypti_ago1k_spline, col.names)

ggplot(aegypti_ago1k_spline , aes(x = x, y = y)) + 
  geom_line(aes(color = sample)) + 
  scale_color_manual(values = c("#3360A9", "#999999"))  + theme_bw()

cells_ago1k  <-split(cells_ago1, cells_ago1$.id)
cells_ago1k  <- lapply(cells_ago1k, function(x) x[order(x$position.x),])
cells_ago1k_spline <-lapply(cells_ago1k, function(x) spline(x$freq_norm))
cells_ago1k_spline <- rbindlist(cells_ago1k_spline, idcol = TRUE)
cells_ago1k_spline$x <- cells_ago1k_spline $x - 101
col.names <- colnames(cells_ago1k_spline)
col.names <- gsub(".id", "sample", col.names)
cells_ago1k_spline <- setNames(cells_ago1k_spline, col.names)

ggplot(cells_ago1k_spline , aes(x = x, y = y)) + 
  geom_line(aes(color = sample)) + 
  scale_color_manual(values = c("#1B0B80", "#999999"))  + theme_bw()


```


##seed search top novel, aegypti specific

```{r}
res <- annotatePeaksWithPatterns(peaks=allGR,
                                  fasta="/Users/kathryn/mirdeep2_master/Aedes-aegypti-LVP_AGWG_CHROMOSOMES_AaegL5.fa",
                                patterns="/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/aegypti_Ago1_master_top.fa",
                                  resize=406,
                                  add5=1,
                                  add3=1,
                                  verbose=TRUE)

#great, has kept all metadata!oh, but only for those tha have matches, will still need to searhc each pattern set, get all res, and rebind together 
res <- res[[1]]
res <- rbindlist(res)

#make sure are only counting pattern matches 1X
#if seqeunce searched AND pattern position are eaxaclty the same, is a duplicate;

res$search_seq_pattern <- paste0(res$seqUnderExtendedPeak, "_", res$siteOfPattern)
nrow(res)
length(unique(res$search_seq_pattern))
# 108465, good, all are unqiue
res$search_seq_pattern <- NULL

#but, could still have dups if extended seq of one peak overlaps with the extended peak of anotheer sequence and then a match is in both; 
#if site of pattern and ex seq is the same, show only the entry with the start of pMatch closest to 0 n
res$group <- paste0(res$SeqExtended, "_", res$siteOfPattern)
length(unique(res$group)) #66228
res$startOfPmatch <- as.numeric(res$startOfPmatch)
res$filt <- duplicated(res$group)
res_Ago1_aegypti_novel <- subset(res, res$filt == FALSE)
nrow(res_Ago1_aegypti_novel)


res_Ago1_aegypti_novel_ctrl <- annotatePeaksWithPatterns(peaks=allctrlGR,
                              fasta="/Users/kathryn/mirdeep2_master/Aedes-aegypti-LVP_AGWG_CHROMOSOMES_AaegL5.fa",
patterns="/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/aegypti_Ago1_master_top.fa",
                                  resize=406,
                                  add5=1,
                                  add3=1,
                                  verbose=TRUE)

res_Ago1_aegypti_novel_ctrl <- res_Ago1_aegypti_novel_ctrl[[1]]
res_Ago1_aegypti_novel_ctrl <- rbindlist(res_Ago1_aegypti_novel_ctrl)

#a <- readDNAStringSet("/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/aegypti_Ago1_master_top.fa")
#b <- readDNAStringSet("/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/top10_Ago1_aegypti_novel_target_FINAL.fa")

#c <- readDNAStringSet("/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/top10_Ago1_aegypti_novel_chimera.fa")
#d <- c(b, c)


#make sure are only counting pattern matches 1X
#if seqeunce searched AND pattern position are eaxaclty the same, is a duplicate;

res_Ago1_aegypti_novel_ctrl$search_seq_pattern <- paste0(res_Ago1_aegypti_novel_ctrl$seqUnderExtendedPeak, "_", res_Ago1_aegypti_novel_ctrl$siteOfPattern)
nrow(res_Ago1_aegypti_novel_ctrl)
length(unique(res_Ago1_aegypti_novel_ctrl$search_seq_pattern))
# 101755, good, all are unqiue
res_Ago1_aegypti_novel_ctrl$search_seq_pattern <- NULL

res_Ago1_aegypti_novel_ctrl$group <- paste0(res_Ago1_aegypti_novel_ctrl$SeqExtended, "_", res_Ago1_aegypti_novel_ctrl$siteOfPattern)
length(unique(res_Ago1_aegypti_novel_ctrl$group)) #62644
res_Ago1_aegypti_novel_ctrl$startOfPmatch <- as.numeric(res_Ago1_aegypti_novel_ctrl$startOfPmatch)
res_Ago1_aegypti_novel_ctrl$filt <- duplicated(res_Ago1_aegypti_novel_ctrl$group)
res_Ago1_aegypti_novel_ctrl <- subset(res_Ago1_aegypti_novel_ctrl, res_Ago1_aegypti_novel_ctrl$filt == FALSE)
nrow(res_Ago1_aegypti_novel_ctrl)

#aae_Ago1_novel_top <- readDNAStringSet("/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/top10_Ago1_aegypti_novel_target_FINAL.fa", format = "fasta")
#aae_Ago1_novel_top%in% res_Ago1_aegypti_novel$Seq
#reverseComplement(aae_Ago1_novel_top) %in% res_Ago1_aegypti_novel$Seq
#reverseComplement(aae_Ago1_novel_top) %in%aae_Ago1_novel_top 

######graph
hist(subset(res_Ago1_aegypti_novel_ctrl$startOfPmatch, res_Ago1_aegypti_novel_ctrl$aegypti_Ago1_BCsample > 0 & res_Ago1_aegypti_novel_ctrl$aegypti_Ago2_BCsample ==0), freq=FALSE, breaks = 200, ylim = range(0, 0.005))
hist(subset(res_Ago1_aegypti_novel$startOfPmatch, res_Ago1_aegypti_novel$aegypti_Ago1_BCsample > 0 & res_Ago1_aegypti_novel$aegypti_Ago2_BCsample ==0), freq=FALSE, breaks = 200, ylim = range(0, 0.005))

Ago1_aegypti_seed_novel_ctrl <- subset(res_Ago1_aegypti_novel_ctrl$startOfPmatch, res_Ago1_aegypti_novel_ctrl$aegypti_Ago1_BCsample > 0 & res_Ago1_aegypti_novel_ctrl$aegypti_Ago2_BCsample ==0)
Ago1_aegypti_seed_novel <- subset(res_Ago1_aegypti_novel$startOfPmatch, res_Ago1_aegypti_novel$aegypti_Ago1_BCsample > 0 & res_Ago1_aegypti_novel$aegypti_Ago2_BCsample ==0)
```

##seed search top novel, Aag2 specific

```{r}
res_Ago1_cells_novel_ctrl <- annotatePeaksWithPatterns(peaks=allctrlGR,
                              fasta="/Users/kathryn/mirdeep2_master/Aedes-aegypti-LVP_AGWG_CHROMOSOMES_AaegL5.fa",
patterns="/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/cells_Ago1_master_top.fa",
                                  resize=406,
                                  add5=1,
                                  add3=1,
                                  verbose=TRUE)

res_Ago1_cells_novel_ctrl <- res_Ago1_cells_novel_ctrl[[1]]
res_Ago1_cells_novel_ctrl <- rbindlist(res_Ago1_cells_novel_ctrl)

#make sure are only counting pattern matches 1X
#if seqeunce searched AND pattern position are eaxaclty the same, is a duplicate;

res_Ago1_cells_novel_ctrl$search_seq_pattern <- paste0(res_Ago1_cells_novel_ctrl$seqUnderExtendedPeak, "_", res_Ago1_cells_novel_ctrl$siteOfPattern)
nrow(res_Ago1_cells_novel_ctrl)
length(unique(res_Ago1_cells_novel_ctrl$search_seq_pattern))
# 101755, good, all are unqiue
res_Ago1_cells_novel_ctrl$search_seq_pattern <- NULL

res_Ago1_cells_novel_ctrl$group <- paste0(res_Ago1_cells_novel_ctrl$SeqExtended, "_", res_Ago1_cells_novel_ctrl$siteOfPattern)
length(unique(res_Ago1_cells_novel_ctrl$group)) #62644
res_Ago1_cells_novel_ctrl$startOfPmatch <- as.numeric(res_Ago1_cells_novel_ctrl$startOfPmatch)
res_Ago1_cells_novel_ctrl$filt <- duplicated(res_Ago1_cells_novel_ctrl$group)
res_Ago1_cells_novel_ctrl <- subset(res_Ago1_cells_novel_ctrl, res_Ago1_cells_novel_ctrl$filt == FALSE)
nrow(res_Ago1_cells_novel_ctrl)

res_Ago1_cells_novel <- annotatePeaksWithPatterns(peaks=allGR,
                              fasta="/Users/kathryn/mirdeep2_master/Aedes-aegypti-LVP_AGWG_CHROMOSOMES_AaegL5.fa",
patterns="/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/cells_Ago1_master_top.fa",
                                  resize=406,
                                  add5=1,
                                  add3=1,
                                  verbose=TRUE)

res_Ago1_cells_novel <- res_Ago1_cells_novel[[1]]
res_Ago1_cells_novel <- rbindlist(res_Ago1_cells_novel)

#make sure are only counting pattern matches 1X
#if seqeunce searched AND pattern position are eaxaclty the same, is a duplicate;

res_Ago1_cells_novel$search_seq_pattern <- paste0(res_Ago1_cells_novel$seqUnderExtendedPeak, "_", res_Ago1_cells_novel$siteOfPattern)
nrow(res_Ago1_cells_novel)
length(unique(res_Ago1_cells_novel$search_seq_pattern))
# 101755, good, all are unqiue
res_Ago1_cells_novel$search_seq_pattern <- NULL

res_Ago1_cells_novel$group <- paste0(res_Ago1_cells_novel$SeqExtended, "_", res_Ago1_cells_novel$siteOfPattern)
length(unique(res_Ago1_cells_novel$group)) #62644
res_Ago1_cells_novel$startOfPmatch <- as.numeric(res_Ago1_cells_novel$startOfPmatch)
res_Ago1_cells_novel$filt <- duplicated(res_Ago1_cells_novel$group)
res_Ago1_cells_novel <- subset(res_Ago1_cells_novel, res_Ago1_cells_novel$filt == FALSE)
nrow(res_Ago1_cells_novel)

hist(subset(res_Ago1_cells_novel_ctrl$startOfPmatch, res_Ago1_cells_novel_ctrl$Aag2_Ago1_BCsample > 0 & res_Ago1_cells_novel_ctrl$Aag2_Ago2_BCsample ==0), freq=FALSE, breaks = 200, ylim = range(0, 0.005))
hist(subset(res_Ago1_cells_novel$startOfPmatch, res_Ago1_cells_novel$Aag2_Ago1_BCsample > 0 & res_Ago1_cells_novel$Aag2_Ago2_BCsample ==0), freq=FALSE, breaks = 200, ylim = range(0, 0.005))

Ago1_Aag2_seed_novel_ctrl <- subset(res_Ago1_cells_novel_ctrl$startOfPmatch, res_Ago1_cells_novel_ctrl$Aag2_Ago1_BCsample> 0 & res_Ago1_cells_novel_ctrl$Aag2_Ago2_BCsample ==0)
Ago1_Aag2_seed_novel <- subset(res_Ago1_cells_novel$startOfPmatch, res_Ago1_cells_novel$Aag2_Ago1_BCsample > 0 & res_Ago1_cells_novel$Aag2_Ago2_BCsample == 0 )
```

##Get peak ranges with patterns to compute read coverage
```{r}
Ago1_aegypti_nranges <- subset(res_Ago1_aegypti_novel, res_Ago1_aegypti_novel$aegypti_Ago1_BCsample > 0 & res_Ago1_aegypti_novel$aegypti_Ago2_BCsample ==0)

Ago1_Aag2_nranges <- subset(res_Ago1_cells_novel, res_Ago1_cells_novel$Aag2_Ago1_BCsample > 0 & res_Ago1_cells_novel$Aag2_Ago2_BCsample == 0 )

Ago1_aegypti_ctrl_nranges <- subset(res_Ago1_aegypti_novel_ctrl, res_Ago1_aegypti_novel_ctrl$aegypti_Ago1_BCsample > 0 & res_Ago1_aegypti_novel_ctrl$aegypti_Ago2_BCsample ==0)

Ago1_Aag2_ctrl_nranges <- subset(res_Ago1_cells_novel_ctrl, res_Ago1_cells_novel_ctrl$Aag2_Ago1_BCsample > 0 & res_Ago1_cells_novel_ctrl$Aag2_Ago2_BCsample == 0 )

##get specific coverage


Ago1_Aag2n_GR <-  GRanges(seqnames = Ago1_Aag2_nranges$seqnames.x, strand = Ago1_Aag2_nranges$strand.x,
                          ranges = IRanges(start = Ago1_Aag2_nranges$start.x, width =Ago1_Aag2_nranges$width.x ))

ToPlot_Aag2_Ago1nalt <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_Aag2Ago1.bw",Ago1_Aag2n_GR ,format = "bigwig")
f <- plotRegion(ToPlot_Aag2_Ago1nalt,outliers = 0.05,groupBy = "Sample")
Aag2_Ago1nalt <- f$data

#Aag2 ctrl ranges 

Ago1_Aag2n_ctrlGR <-  GRanges(seqnames = Ago1_Aag2_ctrl_nranges$seqnames.x, strand = Ago1_Aag2_ctrl_nranges$strand.x,
                          ranges = IRanges(start = Ago1_Aag2_ctrl_nranges$start.x, width =Ago1_Aag2_ctrl_nranges$width.x ))

ToPlot_ctrl_Aag2_Ago1nalt <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_Aag2Ago1.bw",Ago1_Aag2n_ctrlGR ,format = "bigwig")
ctrlAag2_Ago1nalt_peaks <- plotRegion(ToPlot_ctrl_Aag2_Ago1nalt ,outliers = 0.05)
ctrlAag2_Ago1nalt_peaks <- ctrlAag2_Ago1nalt_peaks$data
ctrlAag2_Ago1nalt_peaks$Sample <- c("ctrl_peaks")

Aag2_Ago1nalt <- rbind(ctrlAag2_Ago1nalt_peaks, Aag2_Ago1nalt)

########aegypti novel 
Ago1_aegyptin_GR <-  GRanges(seqnames = Ago1_aegypti_nranges$seqnames.x, strand = Ago1_aegypti_nranges$strand.x,
                          ranges = IRanges(start = Ago1_aegypti_nranges$start.x, width =Ago1_aegypti_nranges$width.x ))

toPlot_aegypti_Ago1nalt <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_aegyptiAgo1.bw",Ago1_aegyptin_GR ,format = "bigwig")
f <- plotRegion(toPlot_aegypti_Ago1nalt,outliers = 0.05,groupBy = "Sample")
aegypti_Ago1nalt <- f$data

#ctrl ranges 
Ago1_aegyptin_ctrlGR <-  GRanges(seqnames = Ago1_aegypti_ctrl_nranges$seqnames.x, strand = Ago1_aegypti_ctrl_nranges$strand.x,
                             ranges = IRanges(start = Ago1_aegypti_ctrl_nranges$start.x, width =Ago1_aegypti_ctrl_nranges$width.x ))


ToPlot_ctrl_aegypti_Ago1nalt <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_aegyptiAgo1.bw",Ago1_aegyptin_ctrlGR,format = "bigwig")
ctrlaegypti_Ago1nalt_peaks <- plotRegion(ToPlot_ctrl_aegypti_Ago1nalt ,outliers = 0.05)
ctrlaegypti_Ago1nalt_peaks <- ctrlaegypti_Ago1nalt_peaks$data
ctrlaegypti_Ago1nalt_peaks$Sample <- c("ctrl_peaks")

aegypti_Ago1nalt <- rbind(ctrlaegypti_Ago1nalt_peaks, aegypti_Ago1nalt)

```

##gather all seed position info and merge with coverage by position, graph

```{r}
nf <- c("Aag2_Ago1nalt", "aegypti_Ago1nalt")
norm_f <- do.call("list",mget(nf))
cov <- lapply(norm_f, function(x) subset(x, x$xIndex >= 1400 & x$xIndex <=1600))
covdf <- rbindlist(cov, idcol = TRUE)
covdf$xIndex <- covdf$xIndex - 1500
col.names <- colnames(covdf)
col.names <- gsub(".id", "ranges", col.names)
col.names <- gsub("xIndex", "position", col.names)
covdf <- setNames(covdf, col.names)

cells_ago1 <- subset(covdf, covdf$ranges=="Aag2_Ago1nalt")
aegypti_ago1 <- subset(covdf, covdf$ranges=="aegypti_Ago1nalt")

cells_ago1$merge <- ifelse(cells_ago1$Sample=="cat_Aag2Ago1.bw" , paste0(cells_ago1$position, "Ago1_Aag2_seed_novel"), paste0(cells_ago1$position, "Ago1_Aag2_seed_novel_ctrl"))

aegypti_ago1$merge <- ifelse(aegypti_ago1$Sample=="cat_aegyptiAgo1.bw" , paste0(aegypti_ago1$position, "Ago1_aegypti_seed_novel"), paste0(aegypti_ago1$position, "Ago1_aegypti_seed_novel_ctrl"))

#make list of pat matches and then count all together for graphing 
Ago1novel <- grep("_seed_novel", names(.GlobalEnv),value=TRUE)
Ago1novel <- grep("Ago1_", Ago1novel, value=TRUE)
Ago1novel <- do.call("list",mget(Ago1novel))
Ago1novel <- lapply(Ago1novel, as.data.frame)
Ago1novel <- lapply(Ago1novel, setNames, nm="position")
Ago1novel_counts <- lapply(Ago1novel, function(x) x %>% group_by(position) %>% summarize(count=n()))
Ago1novel_counts <- lapply(Ago1novel_counts, function(x) subset(x, x$position <= 300 & x$position >=100))

#make master df by sample name
Ago1noveldf <- rbindlist(Ago1novel_counts, idcol = TRUE)
Ago1noveldf$position <- Ago1noveldf$position - 200
Aag2Ago1n <- subset(Ago1noveldf, Ago1noveldf$.id=="Ago1_Aag2_seed_novel_ctrl" | Ago1noveldf$.id=="Ago1_Aag2_seed_novel")
Aag2Ago1n$merge <- paste0(Aag2Ago1n$position,Aag2Ago1n$.id)

aegyptiAgo1n <- subset(Ago1noveldf, Ago1noveldf$.id=="Ago1_aegypti_seed_novel_ctrl" | Ago1noveldf$.id=="Ago1_aegypti_seed_novel")
aegyptiAgo1n$merge <- paste0(aegyptiAgo1n$position, aegyptiAgo1n$.id)

#merge 
aegypti_ago1 <- merge(aegyptiAgo1n, aegypti_ago1, by="merge")
cells_ago1 <- merge(Aag2Ago1n, cells_ago1, by="merge")

#scale data
aegypti_ago1 <- aegypti_ago1 %>%
  group_by(Sample) %>%
  mutate( counts_norm = (count*Score))

aegypti_ago1 <- aegypti_ago1 %>%
  group_by(Sample) %>%
  mutate(freq_norm = counts_norm/(sum(counts_norm)))

aegypti_ago1n  <-split(aegypti_ago1, aegypti_ago1$.id)
aegypti_ago1n  <- lapply(aegypti_ago1n, function(x) x[order(x$position.x),])
aegypti_ago1n_spline <-lapply(aegypti_ago1n, function(x) spline(x$freq_norm))
aegypti_ago1n_spline <- rbindlist(aegypti_ago1n_spline, idcol = TRUE)
aegypti_ago1n_spline$x <- aegypti_ago1n_spline$x - 101
col.names <- colnames(aegypti_ago1n_spline)
col.names <- gsub(".id", "sample", col.names)
aegypti_ago1n_spline <- setNames(aegypti_ago1n_spline, col.names)

ggplot(aegypti_ago1n_spline , aes(x = x, y = y)) + 
  geom_line(aes(color = sample)) + 
  scale_color_manual(values = c("#3360A9", "#999999")) + theme_bw() 

########
cells_ago1 <- cells_ago1 %>%
  group_by(Sample) %>%
  mutate( counts_norm = (count*Score))

cells_ago1 <-cells_ago1 %>%
  group_by(Sample) %>%
  mutate(freq_norm = counts_norm/(sum(counts_norm)))

cells_ago1n <-split(cells_ago1, cells_ago1$.id)
cells_ago1n <- lapply(cells_ago1n, function(x) x[order(x$position.x),])
cells_ago1n_spline <-lapply(cells_ago1n, function(x) spline(x$freq_norm))
cells_ago1n_spline <- rbindlist(cells_ago1n_spline, idcol = TRUE)
cells_ago1n_spline$x <- cells_ago1n_spline $x - 101
col.names <- colnames(cells_ago1n_spline)
col.names <- gsub(".id", "sample", col.names)
cells_ago1n_spline <- setNames(cells_ago1n_spline, col.names)

ggplot(cells_ago1n_spline , aes(x = x, y = y)) + 
  geom_line(aes(color = sample)) + 
  scale_color_manual(values = c("#1B0B80", "#999999")) + theme_bw()

```

##search for known miRNAs that have more similar expression as novel miRNAs in moth mqosuitoes and cells
```{r}
 
res_Ago1_av_known <- annotatePeaksWithPatterns(peaks=allGR,
                                  fasta="/Users/kathryn/mirdeep2_master/Aedes-aegypti-LVP_AGWG_CHROMOSOMES_AaegL5.fa",
                                patterns="/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/average_known_Ago1_target.fa",
                                  resize=406,
                                  add5=1,
                                  add3=1,
                                  verbose=TRUE)

#great, has kept all metadata!oh, but only for those tha have matches, will still need to searhc each pattern set, get all res, and rebind together 
res_Ago1_av_known <- res_Ago1_av_known[[1]]
res_Ago1_av_known <- rbindlist(res_Ago1_av_known)

#make sure are only counting pattern matches 1X
#if seqeunce searched AND pattern position are eaxaclty the same, is a duplicate;

res_Ago1_av_known$search_seq_pattern <- paste0(res_Ago1_av_known$seqUnderExtendedPeak, "_", res_Ago1_av_known$siteOfPattern)
nrow(res_Ago1_av_known)
length(unique(res_Ago1_av_known$search_seq_pattern))
# 108465, good, all are unqiue
res_Ago1_av_known$search_seq_pattern <- NULL

#but, could still have dups if extended seq of one peak overlaps with the extended peak of anotheer sequence and then a match is in both;
res_Ago1_av_known$group <- paste0(res_Ago1_av_known$SeqExtended, "_", res_Ago1_av_known$siteOfPattern)
length(unique(res_Ago1_av_known$group)) #66228
res_Ago1_av_known$startOfPmatch <- as.numeric(res_Ago1_av_known$startOfPmatch)
res_Ago1_av_known$filt <- duplicated(res_Ago1_av_known$group)
res_Ago1_av_known <- subset(res_Ago1_av_known, res_Ago1_av_known$filt == FALSE)
nrow(res_Ago1_av_known)


res_Ago1_av_known_ctrl <- annotatePeaksWithPatterns(peaks=allctrlGR,
                              fasta="/Users/kathryn/mirdeep2_master/Aedes-aegypti-LVP_AGWG_CHROMOSOMES_AaegL5.fa",
patterns="/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/average_known_Ago1_target.fa",
                                  resize=406,
                                  add5=1,
                                  add3=1,
                                  verbose=TRUE)

res_Ago1_av_known_ctrl <- res_Ago1_av_known_ctrl[[1]]
res_Ago1_av_known_ctrl <- rbindlist(res_Ago1_av_known_ctrl)

#make sure are only counting pattern matches 1X
#if seqeunce searched AND pattern position are eaxaclty the same, is a duplicate;

res_Ago1_av_known_ctrl$search_seq_pattern <- paste0(res_Ago1_av_known_ctrl$seqUnderExtendedPeak, "_", res_Ago1_av_known_ctrl$siteOfPattern)
nrow(res_Ago1_av_known_ctrl)
length(unique(res_Ago1_av_known_ctrl$search_seq_pattern))
# 101755, good, all are unqiue
res_Ago1_av_known_ctrl$search_seq_pattern <- NULL

res_Ago1_av_known_ctrl$group <- paste0(res_Ago1_av_known_ctrl$SeqExtended, "_", res_Ago1_av_known_ctrl$siteOfPattern)
length(unique(res_Ago1_av_known_ctrl$group)) #62644
res_Ago1_av_known_ctrl$startOfPmatch <- as.numeric(res_Ago1_av_known_ctrl$startOfPmatch)
res_Ago1_av_known_ctrl$filt <- duplicated(res_Ago1_av_known_ctrl$group)
res_Ago1_av_known_ctrl <- subset(res_Ago1_av_known_ctrl, res_Ago1_av_known_ctrl$filt == FALSE)
nrow(res_Ago1_av_known_ctrl)

##select pattern matches that meet peak criteria 
res_Ago1_av_cells_known_ctrl <- subset(res_Ago1_av_known_ctrl$startOfPmatch, res_Ago1_av_known_ctrl$Aag2_Ago1_BCsample > 0 & res_Ago1_av_known_ctrl$Aag2_Ago2_BCsample == 0)
res_Ago1_av_cells_known <- subset(res_Ago1_av_known$startOfPmatch, res_Ago1_av_known$Aag2_Ago1_BCsample > 0 & res_Ago1_av_known$Aag2_Ago2_BCsample == 0)
res_Ago1_av_aegypti_known_ctrl <- subset(res_Ago1_av_known_ctrl$startOfPmatch, res_Ago1_av_known_ctrl$aegypti_Ago1_BCsample > 0 & res_Ago1_av_known_ctrl$aegypti_Ago2_BCsample == 0)
res_Ago1_av_aegypti_known <- subset(res_Ago1_av_known$startOfPmatch, res_Ago1_av_known$aegypti_Ago1_BCsample > 0 & res_Ago1_av_known$aegypti_Ago2_BCsample == 0)

##make GRs by the same criteria to get specific coverage

res_Ago1_av_cells_known_ctrl_pats <- subset(res_Ago1_av_known_ctrl, res_Ago1_av_known_ctrl$Aag2_Ago1_BCsample > 0 & res_Ago1_av_known_ctrl$Aag2_Ago2_BCsample == 0)
res_Ago1_av_cells_known_pats <- subset(res_Ago1_av_known, res_Ago1_av_known$Aag2_Ago1_BCsample > 0 & res_Ago1_av_known$Aag2_Ago2_BCsample == 0)
res_Ago1_av_aegypti_known_ctrl_pats <- subset(res_Ago1_av_known_ctrl, res_Ago1_av_known_ctrl$aegypti_Ago1_BCsample > 0 & res_Ago1_av_known_ctrl$aegypti_Ago2_BCsample == 0)
res_Ago1_av_aegypti_known_pats <- subset(res_Ago1_av_known, res_Ago1_av_known$aegypti_Ago1_BCsample > 0 & res_Ago1_av_known$aegypti_Ago2_BCsample == 0)

##get specific coverage

Ago1_Aag2_avk_ctrlGR <-  GRanges(seqnames = res_Ago1_av_cells_known_ctrl_pats$seqnames.x, strand = res_Ago1_av_cells_known_ctrl_pats$strand.x,
                          ranges = IRanges(start = res_Ago1_av_cells_known_ctrl_pats$start.x, width =res_Ago1_av_cells_known_ctrl_pats$width.x ))

ToPlot_Aag2_Ago1_avk_ctrl <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_Aag2Ago1.bw",Ago1_Aag2_avk_ctrlGR ,format = "bigwig")
f <- plotRegion(ToPlot_Aag2_Ago1_avk_ctrl,outliers = 0.05,groupBy = "Sample")
Aag2_Ago1_avk_ctrl <- f$data
Aag2_Ago1_avk_ctrl$Sample <- c("ctrl_peaks")

Ago1_Aag2_avk_GR <-  GRanges(seqnames = res_Ago1_av_cells_known_pats$seqnames.x, strand = res_Ago1_av_cells_known_pats$strand.x,
                          ranges = IRanges(start = res_Ago1_av_cells_known_pats$start.x, width =res_Ago1_av_cells_known_pats$width.x ))

ToPlot_Aag2_Ago1_avk <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_Aag2Ago1.bw",Ago1_Aag2_avk_GR ,format = "bigwig")
Aag2_Ago1_avk <- plotRegion(ToPlot_Aag2_Ago1_avk ,outliers = 0.05)
Aag2_Ago1_avk <- Aag2_Ago1_avk$data
Aag2_Ago1_avk <- rbind(Aag2_Ago1_avk_ctrl, Aag2_Ago1_avk)

########aegypti novel 
Ago1_aegypti_avk_ctrlGR <-  GRanges(seqnames = res_Ago1_av_aegypti_known_ctrl_pats$seqnames.x, strand = res_Ago1_av_aegypti_known_ctrl_pats$strand.x,
                          ranges = IRanges(start = res_Ago1_av_aegypti_known_ctrl_pats$start.x, width =res_Ago1_av_aegypti_known_ctrl_pats$width.x ))

toPlot_aegypti_avk_ctrl <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_aegyptiAgo1.bw",Ago1_aegypti_avk_ctrlGR ,format = "bigwig")
aegypti_Ago1_avk_ctrl <- plotRegion(toPlot_aegypti_avk_ctrl,outliers = 0.05,groupBy = "Sample")
aegypti_Ago1_avk_ctrl <- aegypti_Ago1_avk_ctrl$data
aegypti_Ago1_avk_ctrl$Sample <- c("ctrl_peaks")

Ago1_aegypti_avk_GR <-  GRanges(seqnames = res_Ago1_av_aegypti_known_pats$seqnames.x, strand = res_Ago1_av_aegypti_known_pats$strand.x,
                             ranges = IRanges(start = res_Ago1_av_aegypti_known_pats$start.x, width =res_Ago1_av_aegypti_known_pats$width.x ))


ToPlot_ctrl_Ago1_aegypti_avk <- regionPlot(bamFile = "/Users/kathryn/Reprocess_all_paper_datasets/AaegL5_mapped/cat_aegyptiAgo1.bw",Ago1_aegypti_avk_GR,format = "bigwig")
aegypti_Ago1_avk <- plotRegion(ToPlot_ctrl_Ago1_aegypti_avk  ,outliers = 0.05)
aegypti_Ago1_avk <- aegypti_Ago1_avk$data

aegypti_Ago1_avk <- rbind(aegypti_Ago1_avk, aegypti_Ago1_avk_ctrl)

#make list  and then count all together for graphing 
Ago1_av_known <- grep("res_Ago1_av_a", names(.GlobalEnv),value=TRUE)
Ago1_av_known <- c(Ago1_av_known, grep("res_Ago1_av_c", names(.GlobalEnv),value=TRUE))
Ago1_av_known <- grep("pats", Ago1_av_known,value=TRUE, invert = TRUE)

Ago1_av_known  <- do.call("list",mget(Ago1_av_known ))
Ago1_av_known  <- lapply(Ago1_av_known , as.data.frame)
Ago1_av_known  <- lapply(Ago1_av_known , setNames, nm="position")
Ago1_av_known_counts <- lapply(Ago1_av_known , function(x) x %>% group_by(position) %>% summarize(count=n()))
Ago1_av_known_counts <- lapply(Ago1_av_known_counts, function(x) subset(x, x$position <= 300 & x$position >=100))

#make master df by sample name
Ago1avdf <- rbindlist(Ago1_av_known_counts, idcol = TRUE)
Ago1avdf$position <- Ago1avdf$position - 200
Aag2Ago1av <- subset(Ago1avdf, Ago1avdf$.id=="res_Ago1_av_cells_known_ctrl" | Ago1avdf$.id=="res_Ago1_av_cells_known")
Aag2Ago1av$merge <- paste0(Aag2Ago1av$position,Aag2Ago1av$.id)

aegyptiAgo1av <- subset(Ago1avdf, Ago1avdf$.id=="res_Ago1_av_aegypti_known_ctrl" | Ago1avdf$.id=="res_Ago1_av_aegypti_known")
aegyptiAgo1av$merge <- paste0(aegyptiAgo1av$position, aegyptiAgo1av$.id)

##get cov
av <- c("Aag2_Ago1_avk", "aegypti_Ago1_avk")
norm_av <- do.call("list",mget(av))
avcov <- lapply(norm_av, function(x) subset(x, x$xIndex >= 1400 & x$xIndex <=1600))
avcovdf <- rbindlist(avcov, idcol = TRUE)
avcovdf$xIndex <- avcovdf$xIndex - 1500
col.names <- colnames(avcovdf)
col.names <- gsub(".id", "ranges", col.names)
col.names <- gsub("xIndex", "position", col.names)
avcovdf <- setNames(avcovdf, col.names)
avcells_ago1 <- subset(avcovdf, avcovdf$ranges=="Aag2_Ago1_avk")
avaegypti_ago1 <- subset(avcovdf, avcovdf$ranges=="aegypti_Ago1_avk")

avaegypti_ago1$merge <- ifelse(avaegypti_ago1$Sample=="cat_aegyptiAgo1.bw" , paste0(avaegypti_ago1$position, "res_Ago1_av_aegypti_known"), paste0(avaegypti_ago1$position, "res_Ago1_av_aegypti_known_ctrl"))

avcells_ago1$merge <- ifelse(avcells_ago1$Sample=="cat_Aag2Ago1.bw" , paste0(avcells_ago1$position, "res_Ago1_av_cells_known"), paste0(avcells_ago1$position, "res_Ago1_av_cells_known_ctrl"))


#merge 
aegyptiAgo1av <- merge(aegyptiAgo1av, avaegypti_ago1, by="merge")
Aag2Ago1av <- merge(Aag2Ago1av, avcells_ago1, by="merge")


aegypti_ago1av <- aegyptiAgo1av  %>%
  group_by(Sample) %>%
  mutate( counts_norm = (count*Score))

aegypti_ago1av  <- aegypti_ago1av %>%
  group_by(Sample) %>%
  mutate(freq_norm = counts_norm/(sum(counts_norm)))

cells_ago1av <- Aag2Ago1av%>%
  group_by(Sample) %>%
  mutate( counts_norm = (count*Score))

cells_ago1av <-cells_ago1av %>%
  group_by(Sample) %>%
  mutate(freq_norm = counts_norm/(sum(counts_norm)))

aegypti_ago1av  <-split(aegypti_ago1av, aegypti_ago1av$.id)
aegypti_ago1av  <- lapply(aegypti_ago1av, function(x) x[order(x$position.x),])
aegypti_ago1av_spline <-lapply(aegypti_ago1av, function(x) spline(x$freq_norm))
aegypti_ago1av_spline <- rbindlist(aegypti_ago1av_spline, idcol = TRUE)
aegypti_ago1av_spline$x <- aegypti_ago1av_spline$x - 101
col.names <- colnames(aegypti_ago1av_spline)
col.names <- gsub(".id", "sample", col.names)
aegypti_ago1av_spline <- setNames(aegypti_ago1av_spline, col.names)

ggplot(aegypti_ago1av_spline , aes(x = x, y = y)) + 
  geom_line(aes(color = sample)) + 
  scale_color_manual(values = c("#3360A9", "#999999")) + theme_bw()

cells_ago1av <-split(cells_ago1av, cells_ago1av$.id)
cells_ago1av <- lapply(cells_ago1av, function(x) x[order(x$position.x),])
cells_ago1av_spline <-lapply(cells_ago1av, function(x) spline(x$freq_norm))
cells_ago1av_spline <- rbindlist(cells_ago1av_spline, idcol = TRUE)
cells_ago1av_spline$x <- cells_ago1av_spline $x - 101
col.names <- colnames(cells_ago1av_spline)
col.names <- gsub(".id", "sample", col.names)
cells_ago1av_spline <- setNames(cells_ago1av_spline, col.names)


ggplot(cells_ago1av_spline , aes(x = x, y = y)) + 
  geom_line(aes(color = sample)) + 
  scale_color_manual(values = c("#1B0B80", "#999999")) + theme_bw()
