---
title: "Chimera_process_to_upload"
author: "KRG"
date: "7/14/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Rbowtie2)
library(dplyr)
library(stringr)
library(purrr)
library(ggplot2)
library(ggrepel)
```

###first reverse map all small RNAs to reads that did not map to AaegL5 

```{r}
##get unmapped reads from bam
Dir <- "/rugpfs/fs0/rice_lab/scratch/krozen/AaegL5_mapped/unmapped_for_chimera"
bams <- dir(Dir,pattern="*.bam$",full.names = TRUE)
unbam <- function(bams){
  outfa <- paste0(dirname(bams), "/unmapped_", gsub(".bam", ".fa", basename(bams)))
  cmd <- paste0("samtools fasta -f 4 ", bams, " > ", outfa)
   system(cmd,wait = T)
}

for (i in 1:length(bams)) {
  unbam(bams[i])
}


#make indices for mapping
Dir <- "/rugpfs/fs0/rice_lab/scratch/krozen/AaegL5_mapped/unmapped_for_chimera"
unmapped <- dir(Dir,pattern="*.fa$",full.names = TRUE)
bowtie_index <- function(genomeFasta,
                         outFasta=gsub("\\.fa","",genomeFasta)
) {
  require(Rbowtie2)
  if(!dir.exists(outFasta)){
    bowtie2_build(references=genomeFasta, 
                  bt2Index=outFasta)
  }
  
}

for (i in unmapped) {
  bowtie_index(i, i)
}

##align all known and putative small RNAs to reads
bowtie_align <- function(fq,index,sam=gsub("\\.fq|\\.fastq|\\.rm|\\.fa",".sam",index)
) {
  require(Rbowtie2)
  if(!dir.exists(sam)){
    bowtie2(bt2Index = index,
            samOutput = sam,
            seq1 = fq,"--threads 4 -f -L 18 -k 1000000")
    Rsamtools::asBam(sam,gsub("\\.sam","",sam))
  }
  
}

#Set Dir with unmapped fasta
Dir <- "/rugpfs/fs0/rice_lab/scratch/krozen/AaegL5_mapped/unmapped_for_chimera"
ref <- dir(Dir,pattern="*.fa$",full.names = TRUE)
ref <- gsub( ".fa", "", ref)

for (i in ref) {
  bowtie_align("/rugpfs/fs0/rice_lab/scratch/krozen/all_putative_known.fa", i)
}

bam <- dir(Dir,pattern="*.bam$",full.names = TRUE)

bamtobed <- function(file,filtDup=FALSE){
  require(GenomicAlignments)
  require(rtracklayer)
  temp <- readGAlignments(file,
                          param=ScanBamParam(what = "qname"))
  names(temp) <- mcols(temp)$qname
  temp <- granges(temp)
  
  if(filtDup) temp <- temp[!duplicated(temp),]
  export.bed(temp,
             con = gsub("\\.bam",".bed",file))
}

bplapply(bam, bamtobed)

#read in bed and chimera process
Dir <- "/rugpfs/fs0/rice_lab/scratch/krozen/AaegL5_mapped/unmapped_for_chimera/known_novel_sRNA_revmapped"
beds <- dir(Dir, pattern="*.bed$",full.names = TRUE)
chimera <- lapply(beds, function(x) {
   if (!file.size(x) == 0) {
       read.delim(x, header = FALSE, sep = "")
   }
})

names(chimera) <- beds
#use this to get empty beds and remove
t <- lapply(chimera, nrow) 
t <- unlist(t) 
t <- names(t) #get names of beds with entried and read these in 

test <- lapply(t, read.delim, header = FALSE, sep = "")
names(test) <- t 
col.names <- c("rowname", "start", "stop", "name", "score", "strand")
test <- lapply(test, setNames, col.names)

t <- gsub("known_novel_sRNA_revmapped/", "", t)
t <- gsub(".bed", ".fa", t) #so t has names of all fastas to read in 

Dir <- "/rugpfs/fs0/rice_lab/scratch/krozen/AaegL5_mapped/unmapped_for_chimera" 
fa <- dir(Dir, pattern="*.fa$",full.names = TRUE) #actually did this on hpc, but same idea
fasta <- lapply(fa, readDNAStringSet, format = "fasta", nrec = -1L)
names(fasta) <- fa 
fasta <- fasta[names(fasta) %in% t]

fasta <- lapply(fasta, as.data.frame)
fasta <- lapply(fasta, function(x) tibble::rownames_to_column(x))


#merge together read sequence and bed by rowname (read name)
chimera <- map2(fasta, test, ~merge(.x,.y, by = "rowname"))
names(chimera) <- gsub("/rugpfs/fs0/rice_lab/scratch/krozen/AaegL5_mapped/unmapped_for_chimera/unmapped_", "/rugpfs/fs0/rice_lab/scratch/krozen/AaegL5_mapped/unmapped_for_chimera/known_novel_sRNA_revmapped/unmapped_", names(chimera))


#write files to chimera inputs: bed with read name, start, stop, srand, read name, and read sequence
setwd(Dir)
for (x in names(chimera)) {
write.table(chimera[[x]], file=paste0(x,".txt"), sep="\t", quote = FALSE)}
write.table(test, file=".txt", sep="\t")

#process to keep read sequence downstream of small RNA sequence

Dir <- "/Users/kathryn/Reprocess_all_paper_datasets/unmapped_for_chimera/known_novel_sRNA_revmapped/"
files <- dir(Dir, pattern="*fa.txt$",full.names = TRUE)

chimeraProcess <- function(input) {
  
  BR1 <- read.delim(input, header=T)
  BR1 <- BR1[!grepl("AAAAAAAAAAAAAAAAAAAAAAAAA", BR1$name),] #this was high frequency in my libraries removed
  BR1<- BR1[BR1$strand=="+",]
  BR1 <- BR1[duplicated(BR1$rowname)==F,] #come back to this, to remove reads with more than one miRNA mapped 
  BR1$length <- sapply(as.character(BR1$x), nchar)
  BR1$ups.seq <- mapply(substr, x=BR1$x, start=0, stop=BR1$start)
  BR1$dns.seq <- mapply(substr, x=BR1$x, start=BR1$stop, stop=BR1$length)
  outname = paste(input, sep = "")
  outname = gsub(".fa.txt", "_chimera.txt", outname)
  write.table(BR1, outname, quote=F, sep="\t")
}

for (i in 1:length(files)) {
  chimeraProcess(files[i])
}

#reformat output as fasta with only seqeunce downstream of small RNA 
files <- dir(Dir, pattern="*_chimera.txt$",full.names = TRUE)

reformat <- function(input) {
  BR1 <- read.delim(input, header=T)
  BR1$V8 <- paste(BR1$rowname, BR1$name, sep = ";")
  BR1 <- BR1[c(11,10 )]
  BR1$dns.seq <- as.character(BR1$dns.seq)
  BR1$length <- nchar(BR1$dns.seq)
  BR1 <- BR1[BR1$length >= 18,]
  BR1 <- BR1[1:2]
  outname = paste(input, '.fa', sep = "")
  BR2 <- DNAStringSet(BR1$dns.seq, use.names = TRUE)
  names(BR2) <- BR1$V8
  writeXStringSet(BR2, outname, format ="fasta")  
}


for (i in 1:length(files)) {
  reformat(files[i])
}


#now map downstream sequence to AaegL5 genome 

bowtie_align <- function(fq,index,sam=gsub("\\.fq|\\.fastq|\\.rm|\\.fa",".sam",fq)
) {
  require(Rbowtie2)
  if(!dir.exists(sam)){
    bowtie2(bt2Index = index,
            samOutput = sam,
            seq1 = fq,"--threads 4 -f -N 1 -L 18")
    Rsamtools::asBam(sam,gsub("\\.sam","",sam))
  }
  
}

Dir <- "/rugpfs/fs0/rice_lab/scratch/krozen/AaegL5_mapped/unmapped_for_chimera/known_novel_sRNA_revmapped/AaegL5_remap"
fqfiles <- dir(Dir, pattern= "*txt.fa", full.names = TRUE)
bplapply(fqfiles, bowtie_align, index = "/rugpfs/fs0/rice_lab/scratch/krozen/AaegL5_ref/Aedes-aegypti-LVP_AGWG_CHROMOSOMES_AaegL5")

#convert remapped to BED:

bamtobed <- function(file,filtDup=FALSE){
  require(GenomicAlignments)
  require(rtracklayer)
  temp <- readGAlignments(file,
                          param=ScanBamParam(what = "qname"))
  names(temp) <- mcols(temp)$qname
  temp <- granges(temp)
  
  if(filtDup) temp <- temp[!duplicated(temp),]
  export.bed(temp,
             con = gsub("\\.bam",".bed",file))
}


Dir <- "/rugpfs/fs0/rice_lab/scratch/krozen/AaegL5_mapped/unmapped_for_chimera/known_novel_sRNA_revmapped/AaegL5_remap"
mapped <- dir(Dir,pattern="bam$",full.names = TRUE)
bplapply(mapped, bamtobed)

```

##now need to read in remapped chimeric reads, concatenate beds by group and write grouped bed 

```{r}
# moved remapped beds from high performance computer to my machine

Dir <- "/Users/kathryn/Reprocess_all_paper_datasets/unmapped_for_chimera/known_novel_sRNA_revmapped/AaegL5_remap"
beds <- dir(Dir, pattern="*.bed$",full.names = TRUE)

chimera <- lapply(beds, function(x) {
   if (!file.size(x) == 0) {
       read.delim(x, header = FALSE, sep = "")
   }
})

names(chimera) <- beds
#use this to remove beds where nothing remapped
t <- lapply(chimera, nrow) 
t <- unlist(t) 
t <- names(t) #get names and read these in 

chimera <- lapply(t, read.delim, header = FALSE, sep = "")
names(chimera ) <- t #this worked sunce removed empties, used set diff to confirm that they are empties, yes spot checked 

n <- gsub("/Users/kathryn/Reprocess_all_paper_datasets/unmapped_for_chimera/known_novel_sRNA_revmapped/AaegL5_remap/", "", t)
n <- gsub("_chimera.txt.bed", "", n)
names(chimera) <- n

#get remapped chimeras by antibody/lysate and write out concatenated beds 
aegyptiAgo1 %in% names(chimera)
aegyptiAgo1_chimera_remap <- test[c(aegyptiAgo1)]
aegyptiAgo2_chimera_remap <- test[c(aegyptiAgo2)]
aegyptirIgG_chimera_remap <- test[c(aegyptirIgG)]
aegyptimIgG_chimera_remap <- test[c(aegyptimIgG)]
Aag2Ago1_chimera_remap <- test[c(Aag2Ago1)]
Aag2Ago2_chimera_remap <- test[c(Aag2Ago2)]
Aag2rIgG_chimera_remap <- test[c(Aag2rIgG)]
Aag2mIgG_chimera_remap <- test[c(Aag2mIgG)]

aegyptiAgo1_chimera_remap <- do.call(rbind, aegyptiAgo1_chimera_remap)
aegyptiAgo2_chimera_remap <- do.call(rbind, aegyptiAgo2_chimera_remap)  
aegyptirIgG_chimera_remap <- do.call(rbind, aegyptirIgG_chimera_remap) 
aegyptimIgG_chimera_remap <- do.call(rbind, aegyptimIgG_chimera_remap) 
Aag2Ago1_chimera_remap <- do.call(rbind, Aag2Ago1_chimera_remap) 
Aag2Ago2_chimera_remap <- do.call(rbind, Aag2Ago2_chimera_remap) 
Aag2rIgG_chimera_remap <- do.call(rbind, Aag2rIgG_chimera_remap) 
Aag2mIgG_chimera_remap <- do.call(rbind,Aag2mIgG_chimera_remap) 

h <-grep("_chimera_remap",names(.GlobalEnv),value=TRUE)
h<- do.call("list",mget(h))

for(i in seq_along(h)) {
  write.table(h[i], paste(names(h)[i], ".bed", sep = ""), 
              col.names = FALSE, row.names = FALSE, sep = "\t", quote = FALSE)
}
```

##now need to count # of each small RNA by sample, group by antibody/lysate, and normalize for visualization 

```{r}
##count chimeras by small RNA
chimera_counts <- lapply(chimera, function(x) str_split_fixed(x$V4, ";", 2))
chimera_counts <- lapply(chimera_counts, function(x) x[,2])
chimera_counts <- lapply(chimera_counts, as.data.frame)
chimera_counts <- lapply(chimera_counts, setNames, nm = "smallRNA")
chimera_counts <- lapply(chimera_counts, function(x) x %>% group_by(smallRNA) %>% summarize(count=n()))

chimera_counts <- Reduce(function(x, y) merge(x, y, by = "smallRNA", all = TRUE), chimera_counts)
chimera_counts[is.na(chimera_counts)] <- 0 
col.names <- c("smallRNA", n)
chimera_counts <- setNames(chimera_counts, col.names)

#get groups by antibody/lysate
aegyptiAgo1 <- grep(".*(Ago1.*aegypti|aegypti.*Ago1).*", col.names, value="TRUE")
aegyptiAgo2 <- grep(".*(Ago2.*aegypti|aegypti.*Ago2).*", col.names, value="TRUE")
aegyptirIgG <- grep(".*(rIgG.*aegypti|aegypti.*rIgG).*", col.names, value="TRUE")
aegyptimIgG <- grep(".*(mIgG.*aegypti|aegypti.*mIgG).*", col.names, value="TRUE")
Aag2Ago1 <- grep(".*(Ago1.*Aag2|Aag2.*Ago1).*", col.names, value="TRUE")
Aag2Ago2 <- grep(".*(Ago2.*Aag2|Aag2.*Ago2).*", col.names, value="TRUE")
Aag2rIgG <- grep(".*(rIgG.*Aag2|Aag2.*rIgG).*", col.names, value="TRUE")
Aag2mIgG <- grep(".*(mIgG.*Aag2|Aag2.*mIgG).*", col.names, value="TRUE")

#get total chimera counts for each small RNA by ab/lysate
chimera_counts$aegyptiAgo1 <- rowSums(chimera_counts[aegyptiAgo1])
chimera_counts$aegyptiAgo2 <- rowSums(chimera_counts[aegyptiAgo2])
chimera_counts$aegyptirIgG <- rowSums(chimera_counts[aegyptirIgG])
chimera_counts$aegyptimIgG <- rowSums(chimera_counts[aegyptimIgG])
chimera_counts$Aag2Ago1 <- rowSums(chimera_counts[Aag2Ago1])
chimera_counts$Aag2Ago2 <- rowSums(chimera_counts[Aag2Ago2])
chimera_counts$Aag2rIgG <- rowSums(chimera_counts[Aag2rIgG])
chimera_counts$Aag2mIgG <- rowSums(chimera_counts[Aag2mIgG])

#get biological complexity (BC; replicates) for each small RNA by ab/lysate
chimera_counts$aegypti_Ago1_BC <- apply(chimera_counts[aegyptiAgo1], 1, function(x) length(which(x>0)))
chimera_counts$aegypti_Ago2_BC <- apply(chimera_counts[aegyptiAgo2], 1, function(x) length(which(x>0)))
chimera_counts$aegypti_rIgG_BC <- apply(chimera_counts[aegyptirIgG], 1, function(x) length(which(x>0)))
chimera_counts$aegypti_mIgG_BC <- apply(chimera_counts[aegyptimIgG], 1, function(x) length(which(x>0)))
chimera_counts$Aag2_Ago1_BC <- apply(chimera_counts[Aag2Ago1], 1, function(x) length(which(x>0)))
chimera_counts$Aag2_Ago2_BC <- apply(chimera_counts[Aag2Ago2], 1, function(x) length(which(x>0)))
chimera_counts$Aag2_rIgG_BC <- apply(chimera_counts[Aag2rIgG], 1, function(x) length(which(x>0)))
chimera_counts$Aag2_mIgG_BC <- apply(chimera_counts[Aag2mIgG], 1, function(x) length(which(x>0)))

#get input reads (unmapped, before any processing) by group to normalize 
Dir <- "/Users/kathryn/Reprocess_all_paper_datasets/unmapped_for_chimera"
ref <- dir(Dir,pattern="*.fa$",full.names = TRUE)

A <- grep(".*(Ago1.*aegypti|aegypti.*Ago1).*", ref, value="TRUE")
B <- grep(".*(Ago2.*aegypti|aegypti.*Ago2).*", ref, value="TRUE")
C <- grep(".*(rIgG.*aegypti|aegypti.*rIgG).*", ref, value="TRUE")
D <- grep(".*(mIgG.*aegypti|aegypti.*mIgG).*", ref, value="TRUE")
E <- grep(".*(Ago1.*Aag2|Aag2.*Ago1).*", ref, value="TRUE")
G <- grep(".*(Ago2.*Aag2|Aag2.*Ago2).*", ref, value="TRUE")
H <- grep(".*(rIgG.*Aag2|Aag2.*rIgG).*", ref, value="TRUE")
I <- grep(".*(mIgG.*Aag2|Aag2.*mIgG).*", ref, value="TRUE")


fastaProcessmod <- function(input) { 
  require(Biostrings)
  fa <- lapply(input, readDNAStringSet, format = "fasta", nrec = -1L) 
 t <- lapply(fa, length)
 t <- sum(unlist(t))
 print(t)

}

fastaProcessmod(A) #1839545
fastaProcessmod(B) #762423
fastaProcessmod(C) #745977
fastaProcessmod(D) #402387
fastaProcessmod(E) #3368128
fastaProcessmod(G) #2964194
fastaProcessmod(H) #2632945
fastaProcessmod(I) #1423698

#add pseudocount so can calc log2FC 
chimera_counts$aegyptiAgo1_pseudocount <- rowSums(chimera_counts[aegyptiAgo1]) + 1
chimera_counts$aegyptiAgo2_pseudocount<- rowSums(chimera_counts[aegyptiAgo2])+1
chimera_counts$aegyptirIgG_pseudocount <- rowSums(chimera_counts[aegyptirIgG])+1
chimera_counts$aegyptimIgG_pseudocount <- rowSums(chimera_counts[aegyptimIgG])+1
chimera_counts$Aag2Ago1_pseudocount <- rowSums(chimera_counts[Aag2Ago1])+1
chimera_counts$Aag2Ago2_pseudocount <- rowSums(chimera_counts[Aag2Ago2])+1
chimera_counts$Aag2rIgG_pseudocount <- rowSums(chimera_counts[Aag2rIgG])+1
chimera_counts$Aag2mIgG_pseudocount <- rowSums(chimera_counts[Aag2mIgG])+1

#normalize to input reads, unmapped fasta 
chimera_counts$aegyptiAgo1_norm <- ((chimera_counts$aegyptiAgo1_pseudocount)/1839545)*1E6
chimera_counts$aegyptiAgo2_norm <- ((chimera_counts$aegyptiAgo2_pseudocount)/762423)*1E6
chimera_counts$aegyptirIgG_norm <- ((chimera_counts$aegyptirIgG_pseudocount)/745977)*1E6
chimera_counts$aegyptimIgG_norm <- ((chimera_counts$aegyptimIgG_pseudocount)/402387)*1E6
chimera_counts$Aag2Ago1_norm <- ((chimera_counts$Aag2Ago1_pseudocount)/3368128)*1E6
chimera_counts$Aag2Ago2_norm <- ((chimera_counts$Aag2Ago2_pseudocount)/2964194)*1E6
chimera_counts$Aag2rIgG_norm <- ((chimera_counts$Aag2rIgG_pseudocount)/2632945)*1E6
chimera_counts$Aag2mIgG_norm <- ((chimera_counts$Aag2mIgG_pseudocount)/1423698)*1E6


chimera_counts$log2aegyptiAgo1_norm_chimera <- log2(chimera_counts$aegyptiAgo1_norm)
chimera_counts$log2aegyptiAgo2_norm_chimera <- log2(chimera_counts$aegyptiAgo2_norm)
chimera_counts$log2aegyptirIgG_norm_chimera <- log2(chimera_counts$aegyptirIgG_norm)
chimera_counts$log2aegyptimIgG_norm_chimera <- log2(chimera_counts$aegyptimIgG_norm)
chimera_counts$log2Aag2Ago2_norm_chimera <- log2(chimera_counts$Aag2Ago2_norm)
chimera_counts$log2Aag2Ago1_norm_chimera <- log2(chimera_counts$Aag2Ago1_norm)
chimera_counts$log2Aag2rIgG_norm_chimera <- log2(chimera_counts$Aag2rIgG_norm)
chimera_counts$log2Aag2mIgG_norm_chimera <- log2(chimera_counts$Aag2mIgG_norm)


chimera_counts$log2FCAag2Ago1overAgo2_chimera <- log2(chimera_counts$Aag2Ago1_norm) - log2(chimera_counts$Aag2Ago2_norm)
chimera_counts$log2FCaegyptiAgo1overAgo2_chimera <- log2(chimera_counts$aegyptiAgo1_norm) - log2(chimera_counts$aegyptiAgo2_norm)

```



