---
title: "goseq_heatmap_Ago2_perfect_vs_imperfect"
author: "KRG"
date: "7/17/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(magrittr)
library(pheatmap)
library(goseq)
library(GenomicFeatures)
```

##Read in eggNOG-mapper v2 tsv, mapping geneIDs to annotations and pathways 

```{r }
tsv <- read.delim("/Users/kathryn/Reprocess_all_paper_datasets/job_MM_e0ep_7ci_annotations.tsv", header = FALSE, sep ="\t",comment.char = "",skip = 4)
tsvv <- tsv %>% dplyr::select(V1,V7) %>% mutate(Gene=gsub("-..","",V1))
tsvv %>% group_by(Gene) %>% mutate(newGO=paste(V7,sep=","))
tsvv <- tsvv[!duplicated(tsvv$Gene),]
myVec <- tsvv$V7
names(myVec) <- tsvv$Gene
myVec <- strsplit(as.vector(myVec),",")
names(myVec) <- paste0(tsvv$Gene,"_ID")
jsjs <- unlist(myVec)
myGOFrame <- data.frame(ID=gsub("_ID.*","",names(jsjs)),GO=jsjs,row.names = NULL)
myss <- as.vector(myGOFrame[,1])

## get transcript lengths 
gtffile <- "/Users/kathryn/Aedes-aegypti-LVP_AGWG_BASEFEATURES_AaegL5.2.gtf"

TxDb <- makeTxDbFromGFF(gtffile)

myLens <- transcriptLengths(TxDb) %>% group_by(gene_id) %>% summarise(gene_len=median(tx_len))
myLens2 <- as.vector(myLens$gene_len)
names(myLens2) <- myLens$gene_id

##get gene lists
all_with_pats <- read.delim("/Users/kathryn/Reprocess_all_paper_datasets/Supp_tables/TableS3_final.txt", header = TRUE, sep="\t")
all_merge_aegypti_Ago2 <- subset(all_with_pats, all_with_pats$aegypti_Ago2_BCsample > 0)
aegypti_perf <- all_merge_aegypti_Ago2[grepl("novel|aae", all_merge_aegypti_Ago2$Pperfect_18mer_target),]
aegypti_perf <- aegypti_perf[grepl("common-Ago2|aegypti-Ago2", aegypti_perf$Pperfect_18mer_target),]

Aag2_perf <- subset(all_with_pats, all_with_pats$Aag2_Ago2_BCsample > 0)
Aag2_perf <-Aag2_perf[grepl("novel|aae", Aag2_perf$Pperfect_18mer_target),]
Aag2_perf <-Aag2_perf[grepl("common-Ago2|Aag2-Ago2", Aag2_perf$Pperfect_18mer_target),]


aegypti_pat_chim <- subset(all_with_pats, all_with_pats$aegypti_Ago2_BCsample > 0)
aegypti_pat_chim <- rbind( aegypti_pat_chim[grepl("Ago2_both|Ago2_aegypti", aegypti_pat_chim$top_smallRNA_chimera),], aegypti_pat_chim[grepl("Ago2_both|Ago2_aegypti", aegypti_pat_chim$Ptop_smallRNA),])
aegypti_pat_chim<- aegypti_pat_chim[!aegypti_pat_chim$peakID %in% aegypti_perf$peakID,]

Aag2_pat_chim <- subset(all_with_pats, all_with_pats$Aag2_Ago2_BCsample > 0)
Aag2_pat_chim <- rbind(Aag2_pat_chim[grepl("Ago2_both|Ago2_cells", Aag2_pat_chim$top_smallRNA_chimera),], Aag2_pat_chim[grepl("Ago2_both|Ago2_cells", Aag2_pat_chim$Ptop_smallRNA),])
Aag2_pat_chim<- Aag2_pat_chim[!Aag2_pat_chim$peakID %in% Aag2_perf$peakID,]

##goseq for each geneset
aegypti_perf_Ago2 <- aegypti_perf$geneId
aegypti_perf_Ago2 <- aegypti_perf_Ago2[!is.na(aegypti_perf_Ago2)] 
g <- myss %in% as.vector(aegypti_perf_Ago2)+0
names(g) <- myss
g <- g[!duplicated(names(g))]
length(which(g==1)) #12

xcscswcd <- nullp(g,NULL,NULL,bias.data = myLens2[match(names(g),names(myLens2),incomparables = 0)])
GOaegypti_Ago2_perf <- goseq(xcscswcd,NULL,NULL,gene2cat = myGOFrame)

aegypti_pat_chim_Ago2 <- aegypti_pat_chim$geneId
aegypti_pat_chim_Ago2 <- aegypti_pat_chim_Ago2[!is.na(aegypti_pat_chim_Ago2)] 
h <- myss %in% as.vector(aegypti_pat_chim_Ago2)+0
names(h) <- myss
h <- h[!duplicated(names(h))]
length(which(h==1)) #1154

xcscswcd <- nullp(h,NULL,NULL,bias.data = myLens2[match(names(h),names(myLens2),incomparables = 0)])
GOaegypti_Ago2_pat_chim <- goseq(xcscswcd,NULL,NULL,gene2cat = myGOFrame)

Aag2_perf_Ago2 <-Aag2_perf$geneId
Aag2_perf_Ago2 <- Aag2_perf_Ago2[!is.na(Aag2_perf_Ago2)] 
j <- myss %in% as.vector(Aag2_perf_Ago2)+0
names(j) <- myss
j <- j[!duplicated(names(j))]
length(which(j==1))

xcscswcd <- nullp(j,NULL,NULL,bias.data = myLens2[match(names(j),names(myLens2),incomparables = 0)])
GOAag2_Ago2_perf <- goseq(xcscswcd,NULL,NULL,gene2cat = myGOFrame)

Aag2_pat_chim_Ago2 <- Aag2_pat_chim$geneId
Aag2_pat_chim_Ago2 <- Aag2_pat_chim_Ago2[!is.na(Aag2_pat_chim_Ago2)] 
k<- myss %in% as.vector(Aag2_pat_chim_Ago2)+0
names(k) <- myss
k <- k[!duplicated(names(k))]
length(which(k==1)) 

xcscswcd <- nullp(k,NULL,NULL,bias.data = myLens2[match(names(k),names(myLens2),incomparables = 0)])
GOAag2_Ago2_pat_chim <- goseq(xcscswcd,NULL,NULL,gene2cat = myGOFrame)

#merge all tables together
go_merge <- merge(GOAag2_Ago2_pat_chim, GOAag2_Ago2_perf, by="category")
go_merge <- go_merge[,c(1,6:7,2:5, 8:11)]
col.names <- colnames(go_merge)
col.names <- c(gsub(".x", "", col.names[1:3]), gsub(".x", "_cells_pat_chim", col.names[4:7]),gsub(".y", "_cells_perf", col.names[8:11])) 
go_merge <- setNames(go_merge, col.names)
go_merge2 <- merge(GOaegypti_Ago2_pat_chim, GOaegypti_Ago2_perf, by="category")
go_merge2 <- go_merge2[,c(1,2:5, 8:11)]
col.names <- colnames(go_merge2)
col.names <- c(col.names[1], gsub(".x", "_aegypti_pat_chim", col.names[2:5]),gsub(".y", "_aegypti_perf", col.names[6:9])) 
go_merge2 <- setNames(go_merge2, col.names)
go_merge <- merge(go_merge, go_merge2, by="category")
#write.table(go_merge, "/Users/kathryn/Reprocess_all_paper_datasets/Supp_tables/Table_S7.txt", col.names = TRUE, row.names = FALSE, sep ="\t", quote = FALSE)
#select significant
go_merge <- subset(go_merge, go_merge$over_represented_pvalue_cells_pat_chim <0.05 | go_merge$under_represented_pvalue_cells_pat_chim < 0.05 | go_merge$over_represented_pvalue_cells_perf < 0.05 | go_merge$under_represented_pvalue_cells_perf <0.05 | go_merge$over_represented_pvalue_aegypti_pat_chim < 0.05 | go_merge$under_represented_pvalue_aegypti_pat_chim <0.05 | go_merge$over_represented_pvalue_aegypti_perf <0.05 | go_merge$under_represented_pvalue_aegypti_perf <0.05)

#just keep overrepresented
go_merge_over_pval <- go_merge[,c(1:2, 4,8,12,16)]
row.names(go_merge_over_pval) <- go_merge_over_pval[,c(2)]
go_merge_over_pval <- subset(go_merge_over_pval, go_merge_over_pval$over_represented_pvalue_cells_pat_chim <0.01 | go_merge_over_pval$over_represented_pvalue_cells_perf <0.01 | go_merge_over_pval$over_represented_pvalue_aegypti_pat_chim <0.01 | go_merge_over_pval$over_represented_pvalue_aegypti_perf <0.01)
#write.table(go_merge_over_pval[,1], "/Users/kathryn/Reprocess_all_paper_datasets/GO_analysis/GOsem_up.txt", col.names = FALSE, row.names = FALSE, sep = "\t", quote = FALSE)
go_merge_over_pval <- as.matrix(go_merge_over_pval[,c(3:6)] )#take higher significance because many terms 

#simplfy by semantic redundancy, keep the most unique terms (from REVIGO, only BPs)
BP <- read.csv("/Users/kathryn/Reprocess_all_paper_datasets/GO_analysis/REVIGO_FINAL.csv", header = TRUE)
#BP2 <- read.csv("/Users/kathryn/Reprocess_all_paper_datasets/GO_analysis/REVIGO_sig_over_BP_thru250.csv", header = TRUE)
#BP <- rbind(BP, BP2)

BP <- BP[!grepl("null", BP$plot_X),]
go_merge_over_pval <- go_merge_over_pval[row.names(go_merge_over_pval) %in% BP$description,] 

#select 10 most significant terms for each set
go_merge_over_pval_filt <- as.data.frame(go_merge_over_pval) %>% tibble::rownames_to_column("GO") %>% dplyr::top_n(-10, over_represented_pvalue_aegypti_perf)
go_merge_over_pval_filt <- rbind(go_merge_over_pval_filt, as.data.frame(go_merge_over_pval) %>% tibble::rownames_to_column("GO") %>% dplyr::top_n(-10, over_represented_pvalue_aegypti_pat_chim))
go_merge_over_pval_filt <-  rbind(go_merge_over_pval_filt,as.data.frame(go_merge_over_pval) %>% tibble::rownames_to_column("GO") %>% dplyr::top_n(-10, over_represented_pvalue_cells_pat_chim))
go_merge_over_pval_filt <-  rbind(go_merge_over_pval_filt,as.data.frame(go_merge_over_pval) %>% tibble::rownames_to_column("GO") %>% dplyr::top_n(-10, over_represented_pvalue_cells_perf))
topVarGenes <- unique(go_merge_over_pval_filt$GO)

#make metadata for heatmap
my_sample_col <- data.frame(target = c("imperfect", "perfect", "imperfect", "perfect"))
row.names(my_sample_col) <- colnames(go_merge_over_pval)
my_sample_col$lysate <- rep(c("cells", "aegypti"), c(2,2))

my_colour = list(
  lysate = c(cells = "#CC0000", aegypti = "#660000"),
  target = c(imperfect = "gray", perfect = "black"))

pal <- colorRampPalette(c("white", "#ED3B2C"))

#-log10 transform pvalues 
log10gomerge <- -log10(go_merge_over_pval)
#set threshold to seperate less extreme pvals
log10gomerge[log10gomerge >5] <- 5
log10gomerge[log10gomerge < 1.3 ] <- 0

```

##Figure 5M

```{r }
pheatmap(log10gomerge[topVarGenes,], annotation_col = my_sample_col, cutree_rows = 2, cutree_cols = 2, show_colnames = F, border_color = "grey60", treeheight_row = 0, treeheight_col =20, color = pal(255), fontsize_row = 6)
```
