---
title: "Code used for for the manuscript: Argonaute-CLIP delineates versatile, functional RNAi networks in *Aedes aegypti*"

author: "Kathryn Rozen-Gagnon"
date: "8/25/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
###Code used for the analysis of AGO1- and AGO2-CLIP in *Aedes aegypti* whole mosquitoes and Aag2 cells

###This code accompanies the manuscript: Argonaute-CLIP delineates versatile, functional RNAi networks in *Aedes aegypti*, a major vector of human viruses

###All source code is available [here](https://github.com/kathrynrozengagnon/AGOCLIP_2020), and was developed into a freely-available, reproducible, and streamlined R package,  [CLIPflexR](https://kathrynrozengagnon.github.io/CLIPflexR/) that eases CLIP analysis in both model and non-model organisms 

Processing/mapping/peak matrix building code for AGO targets can be seen on my github (there are no graphical outputs): [link](Processing_to_matrix.html)

Click [link](PCA_annotation_fgsea_targetvenns.html) to see:  
PCAs of all unfiltered samples (Figure S3D), filtering of high-confidence AGO targets (data underlying Figures S3A-C), and PCAs of high-confidence targets in cells and whole mosquitoes (Figures 4G and 4J)  
fgsea of high-confidence gene targets associated with AGO1 or AGO2 (Figures 4H, 4I, 4K, 4L and Tables S5 and S6)  
genic annotation of high-confidence targets (Figures 4C-4F)  
Venn diagram of gene targets of AGO1 and AGO2 in cells and whole mosquitoes (Figure S3E) and venn diagram comparing gene targets from [Zhang et al., 2017](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5347622/) (Figure S3F)  

Click [link](mirdeep2_processing_filtering_to_upload.html) to see:  
Processing/miRDeep2 discovery/quantification and filtering to obtain high-confidence novel small RNAs   
Abundance of high-confidence known miRNAs and novel small RNAs in AGO1 and AGO2 (Figures 4A and 4B)  
Length distribution of high-confidence novel small RNAs and known miRNAs associated with AGO1 or AGO2, in cells or whole mosquitoes (Figure S2A and S2B)  
Venn diagrams showing presence of AGO1-associated high-confidence known (Figure S2C) or novel (Figure S2D) miRNAs in cells compared to whole mosquitoes   

Chimera processing to identify reads containing small RNAs and targets can be seen on my github (there are no graphical outputs): [link](chimera_process_to_upload.html)

Click [link](Ago1_known_novel_target_enrichment_in_peaks_to_upload.html) to see:  
Enrichment of abundant known (Figures 5A and S4A) and novel miRNAs (Figures 5B and S4B) in sequences in AGO1 peaks   
Enrichment of known miRNAs that have similar abundances as the most abundant novel miRNAs (Figures S4C and S4D) in sequences in AGO1 peaks  

Click [link](smallRNA_abundance_scatterplots_to_upload.html) to see:  
Quantification of chimera abundance, and comparison of high-confidence known miRNAs and novel small RNAs by overall abundance, or by chimera abundance (Figures 5D, S4F, S5C, and S5E)   
Table S3 formatting, naming of novel small RNAs, and getting the most abundant small RNA families 

Click [link](Ago1_miRNA_target_abundances_pies_TableS4_to_upload.html) to see:  
Scatterplots showing the correlation between miRNA families and their targets, and inset pie charts showing the percentages of high-confidence miRNAs by class (Figure 5C and S4E)  
Table S4 formatting

Click [link](metagene_plots_coverage_chimeras.html) to see:  
Read coverage over 3'UTRs for AGO1  (Figure 5E and S4G) and AGO2 (Figure 6A)  
Frequency of chimeras and predicted 6mers over 3'UTRs for AGO1 (Figures 5F and S4H) and AGO2 (Figures S5D and S5F)

Click [link](Repeat_features_to_upload.html) to see:  
Mapping all samples to repeat features and visualization of repeat feature families with signficantly higher percentages of mapped reads in AGO2 libraries versus paired IgG controls (Figure 6B)

Click to see:  
Quantification of size and strand bias for [*Ae. aegypti*-mapped](overlaps_strand_bias_to_upload.html) (Figures 6E and 6F) or [CFAV-mapped](CFAV_short_reads_size_strand_distribution.html) (Figure 6H) short reads, from AGO1 or AGO2 libraries

Click [link](Ago2_esiRNA_target_enrichment_in_peaks_mod.html) to see:  
Enrichment of abundant novel esiRNA sequences in AGO2 peaks (Figures 6I and S5B)
Histogram showing the number of perfect 18mer targets from the same esiRNA family found on AGO2-bound transcripts (Figure S5G) Putative AGO2 specificity motifs (Figures 6J and S5H)

Click [link](goseq_heatmap_Ago2_perfect_vs_imperfect_to_upload.html) to see:
goseq analysis of biologiical processes regulated by perfect or imperfect targeting of AGO2 in cells and mosquitoes (Figure 6K)

Click [link](RNAseq_target_validation_to_upload.html) to see:  
Selection and normalized expression of mosquito ovarian miRNA targets (Figures 7A-7C) or control miRNA targets (Figures S6A and S6B), or their relative expression in all individual mosquito tissues from [Matthews et al., 2016](https://pubmed.ncbi.nlm.nih.gov/26738925/) (Figure S6C)  
Selection of targets by type for abundant novel mosquito esiRNAs or AGO2-associated known miRNAs, their relative expression in all individual mosquito tissues from Matthews et al., 2016 (Figure S6D), or their normalized expression in ovaries compared to all tissues (Figures 7E and 7F)

Click [link](reprocess_reanalyze_smallRNA_seq_Akbari_to_upload.html) to see:
Reprocessing of small RNA-seq data from [Akbari et al., 2013](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3755910/) to quantify novel esiRNA expression in ovaries (Figure 7D)  
  