---
title: "RNAseq_target_validation"
author: "KRG"
date: "7/30/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(edgeR)
library(dplyr)
library(GSVA)
library(ggplot2)
require(tidyverse)
library(magrittr)
library(GenomicFeatures)
library(reshape2)
library(Biostrings)
```

## Select Ago1 targets for ovary-expressed miRNAs

```{r}
all_with_pats <- read.delim("/Users/kathryn/Reprocess_all_paper_datasets/Supp_tables/TableS3_final_rn2.txt", header = TRUE, sep="\t", stringsAsFactors = FALSE)

#all_merge$filt <- all_merge$peakID %in% aegypti_Ago1_peaks_filt$peakID
all_merge_aegypti_Ago1 <- subset(all_with_pats, all_with_pats$aegypti_Ago1_BCsample > 0)
aegypti_Ago1_3UTR <- all_merge_aegypti_Ago1[grepl( "3' UTR", all_merge_aegypti_Ago1$annotation),]

##miR989
all_3UTR_989 <- rbind(aegypti_Ago1_3UTR[grepl("miR-989x", aegypti_Ago1_3UTR$aegyptiAgo1_chimera),], aegypti_Ago1_3UTR[grepl("miR-989", aegypti_Ago1_3UTR$PsmallRNA_family),])

##miR996
all_3UTR_996 <- rbind(aegypti_Ago1_3UTR[grepl("miR-996x", aegypti_Ago1_3UTR$aegyptiAgo1_chimera),], aegypti_Ago1_3UTR[grepl("miR-996", aegypti_Ago1_3UTR$PsmallRNA_family),])

##miR2946
all_3UTR_2946 <- rbind(aegypti_Ago1_3UTR[grepl("miR-2946x", aegypti_Ago1_3UTR$aegyptiAgo1_chimera),], aegypti_Ago1_3UTR[grepl("miR-2941", aegypti_Ago1_3UTR$PsmallRNA_family),]) #2946 shares seed with 2941, family name is 2941

#write out gene names
h <-grep("all_3UTR",names(.GlobalEnv),value=TRUE)
h<- do.call("list",mget(h))

h <- lapply(h, function(x) x[,c("geneId", "transcriptId")])
hu <- lapply(h, unique)

#run this to write out gene lists:
#for(i in seq_along(hu)) {
#  write.table(hu[i], paste0("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Ago1_lists_FINAL/", names(hu)[i], ".txt"), 
#              col.names = TRUE, row.names = FALSE, sep = "\t", quote = FALSE)
#}

#get diff genes as control from Matthews et al. ; all data used here are available at: https://github.com/VosshallLab/Aedes_neurotranscriptome

diff_genes <- read.delim("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Matthews_S6.txt", header = TRUE, sep = "\t")

Fe_Ov_all <- grep("Fe_Ov", colnames(diff_genes), value = TRUE)
Fe_all <- grep("Fe", colnames(diff_genes), value = TRUE)

diff_genes$Ov_av_all <- rowMeans(diff_genes[Fe_Ov_all])
diff_genes$all_Fe_notOv <- rowMeans(diff_genes[Fe_all])

diff_genes$Ov_hi <- diff_genes$Ov_av_all/diff_genes$all_Fe
diff_genes$Ov_lo <- diff_genes$all_Fe/diff_genes$Ov_av_all

Ov_hi <- diff_genes %>% top_n(50 , Ov_hi) %>% dplyr::select(gene)
Ov_lo <- diff_genes %>% top_n(-50 , Ov_hi) %>% dplyr::select(gene)

h <-grep("Ov",names(.GlobalEnv),value=TRUE)
h <- grep("lo|hi", h, value = TRUE)
h<- do.call("list",mget(h))
hu <- lapply(h, unique)
#for(i in seq_along(hu)) {
#  write.table(hu[i], #paste0("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Ago1_lists_FI#NAL/", names(hu)[i], ".txt"), 
#              col.names = TRUE, row.names = FALSE, sep = "\t", quote = FALSE)
#}

```

##GSVA for Ago1 ovary-specific miRNA targets (Figure S6A)

```{r }
##get counts from Matthews et al., available at: https://github.com/VosshallLab/Aedes_neurotranscriptome

PECounts <- read.delim("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Aedes_neurotranscriptome-home-newDESEQ2/raw_gene_counts/PE_featurecounts.txt",sep = "\t",skip=1)
SECounts <- read.delim("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Aedes_neurotranscriptome-home-newDESEQ2/raw_gene_counts/SE_featurecounts.txt",sep = "\t",skip=1)
PECounts2 <- PECounts[,grepl("bam",colnames(PECounts))] %>% as.matrix %>% set_colnames(gsub("_\\..*","",colnames(.))) %>% set_rownames(PECounts$Geneid)
SECounts2 <- SECounts[,grepl("bam",colnames(SECounts))] %>% as.matrix %>% set_colnames(gsub("_merged.bam|bams\\.|Aligned.*|_SE","",colnames(.))) %>% set_rownames(SECounts$Geneid)
annotation <- read.delim("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Aedes_neurotranscriptome-home-newDESEQ2/annotations/annotation_with_orthodb-20150730.csv",sep=",")

myAgo1List <- dir("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Ago1_lists_FINAL/",full.names = TRUE) %>% lapply(function(x)read.delim(x,header=TRUE)[,1] %>% as.vector)

#map to their geneIDs
Ago1setList <- lapply(myAgo1List,function(x) unique(annotation[annotation[,2] %in% x,1]) %>% na.omit %>% as.vector)
names(Ago1setList) <- dir("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Ago1_lists_FINAL/") %>% gsub("\\.txt","",.)

SECounts2 <- SECounts2[,!colnames(SECounts2) %in% c("Fe_An_BF_1","Fe_An_SF_5")]
myNewCounts <- merge(SECounts2,PECounts2,by=0)

myNewCounts <- myNewCounts[,!colnames(myNewCounts) %in% "Fe_Br_SF_2"]
metaData<- data.frame(Group=gsub("_\\d$","",colnames(myNewCounts)),row.names = colnames(myNewCounts))
metaData2 <- data.frame(Group=gsub("_\\d$","",colnames(myNewCounts)),
                        Paired=ifelse(colnames(myNewCounts) %in% colnames(PECounts2),"Paired","Single"),
                        row.names = colnames(myNewCounts))

myNewCounts2 <- myNewCounts[,-1]
rownames(myNewCounts2) <- myNewCounts[,1]
myCounts2 <- edgeR::cpm(myNewCounts2)
require(GSVA)

Fe <- grep("Fe", colnames(myCounts2), value = TRUE)
mhCountsAgo1 <- gsva(myCounts2[,colnames(myCounts2) %in% Fe],gset.idx.list = Ago1setList[grepl("UTR|Ov", names(Ago1setList))],kcdf=c("Poisson"))

#visualize with heatmap
colors <- metaData2
colors <- as.data.frame(unique(colors$Group))
colors <- setNames(colors, c("Group"))
colors$color <- ifelse(grepl("Ov_O", colors$Group), paste0("lightslateblue"), paste0(colors$color))
colors$color <- ifelse(grepl("Ov_SF", colors$Group), paste0("plum"), paste0(colors$color))
mycolors <- as.character(colors$color) 
names(mycolors) <- colors$Group
mycolors <- list(Group = mycolors)
mycolors$Paired <- c("gray50", "black")
names(mycolors$Paired) <- c("Paired", "Single")

paletteLength <- 50
palAgo1 <- colorRampPalette(c( "#08306B","white", "darkorange"))(paletteLength)
myBreaks <- c(seq(min(mhCountsAgo1), 0, length.out=ceiling(paletteLength/2) + 1), seq(max(mhCountsAgo1)/paletteLength, max(mhCountsAgo1), length.out=floor(paletteLength/2)))

pheatmap:::pheatmap(mhCountsAgo1,
                    annotation_col = metaData2,
                     color = palAgo1, breaks = myBreaks,
                    cluster_cols = TRUE,cluster_rows = TRUE, scale="none", fontsize_col = 6, annotation_colors = mycolors, treeheight_row = 20, fontsize = 6, cellheight = 25, cellwidth = 6)

```

##eCDFs of miRNA target expression, Figure 6A-C

```{r}
#need to normalize by transcript length for this analysis, use gtf with Matthews et al. gene ids - are not vectorbase IDs, available at: https://github.com/VosshallLab/Aedes_neurotranscriptome

gtffile <- "/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Aedes_neurotranscriptome-home-newDESEQ2/annotations/STAR_featureCount.gtf"

TxDb <- makeTxDbFromGFF(gtffile)

myLens <- transcriptLengths(TxDb) %>% group_by(gene_id) %>% summarise(gene_len=median(tx_len))

myNewCounts3 <- merge(myNewCounts2, myLens, by.x=0, by.y= "gene_id", all.x=TRUE)
row.names(myNewCounts3) <- myNewCounts3$Row.names

d <- DGEList(counts = myNewCounts3[,c(2:130)])
d$genes$Length <- c(myNewCounts3$gene_len)
myCountstest <- edgeR::rpkm(d)
myCountstest <- as.data.frame(myCountstest)


Fe_all <- grep("Fe", colnames(myCountstest), value = TRUE)
Fe_Ov <- grep("Ov", colnames(myCountstest), value = TRUE)
Fe_Ov_SF <- grep("SF", Fe_Ov, value = TRUE)
Fe_Ov_O <- grep("SF", Fe_Ov, value = TRUE, invert = TRUE)

myCountstest$Fe_Ov_SF_av <- rowMeans(myCountstest[Fe_Ov_SF])
myCountstest$Fe_Ov_O_av <- rowMeans(myCountstest[Fe_Ov_O])
myCountstest$Fe_mean <- rowMeans(myCountstest[Fe_all])

##miR-989
test989 <- myCountstest[row.names(myCountstest) %in% Ago1setList$all_3UTR_989,]
test989<- test989[grepl("mean|av", colnames(test989))]
test989<- melt(test989)
ggplot(test989, aes(log2(value), colour = variable)) + stat_ecdf(geom = "point", size=1) + scale_colour_manual(values=c("plum", "lightslateblue","gray")) + theme_bw() 

ks.test(subset(test989$value, test989$variable=="Fe_Ov_O_av"), subset(test989$value, test989$variable=="Fe_mean"), alternative = c("greater"))

ks.test(subset(test989$value, test989$variable=="Fe_Ov_SF_av"), subset(test989$value, test989$variable=="Fe_mean"), alternative = c("greater"))


##miR-996
test996 <- myCountstest[row.names(myCountstest) %in% Ago1setList$all_3UTR_996,]
test996<- test996[grepl("mean|av", colnames(test996))]
test996<- melt(test996)
ggplot(test996, aes(log2(value), colour = variable)) + stat_ecdf(geom = "point", size=1) + scale_colour_manual(values=c("plum", "lightslateblue","gray")) + theme_bw() 

ks.test(subset(test996$value, test996$variable=="Fe_Ov_O_av"), subset(test996$value, test996$variable=="Fe_mean"), alternative = c("greater"))

ks.test(subset(test996$value, test996$variable=="Fe_Ov_SF_av"), subset(test996$value, test996$variable=="Fe_mean"), alternative = c("greater"))

##miR-2946
test2946 <- myCountstest[row.names(myCountstest) %in% Ago1setList$all_3UTR_2946,]
test2946<- test2946[grepl("mean|av", colnames(test2946))]
test2946<- melt(test2946)
ggplot(test2946, aes(log2(value), colour = variable)) + stat_ecdf(geom = "point", size=1) + scale_colour_manual(values=c("plum", "lightslateblue","gray")) + theme_bw() 

ks.test(subset(test2946$value, test2946$variable=="Fe_Ov_SF_av"), subset(test2946$value, test2946$variable=="Fe_mean"), alternative = c("greater"))

ks.test(subset(test2946$value, test2946$variable=="Fe_Ov_O_av"), subset(test2946$value, test2946$variable=="Fe_mean"), alternative = c("greater"))

```

## Select Ago2 targets for ovary-expressed esiRNAs

```{r}
#get 3'UTRs and Ago2-supported peaks
all_merge_aegypti_Ago2 <- subset(all_with_pats, all_with_pats$aegypti_Ago2_BCsample > 0)
aegypti_Ago2_3UTR <- all_merge_aegypti_Ago2[grepl( "3' UTR", all_merge_aegypti_Ago2$annotation),]

#perfect targets (all, doesn't need to be 3'UTR)
aegypti_perf <- all_merge_aegypti_Ago2[grepl("novel|aae", all_merge_aegypti_Ago2$Pperfect_18mer_target),]
aegypti_perf <- aegypti_perf[grepl("common-Ago2|aegypti-Ago2", aegypti_perf$Pperfect_18mer_target),]

#chimera 3'UTR targets
aegypti_chim_Ago2 <- aegypti_Ago2_3UTR[grepl("Ago2_both|Ago2_aegypti", aegypti_Ago2_3UTR$top_smallRNA_chimera),]
aegypti_chim_Ago2 <- aegypti_chim_Ago2[!aegypti_chim_Ago2$peakID %in% aegypti_perf$peakID,]

#6mer 3'UTR targets
aegypti_6mer_Ago2 <- aegypti_Ago2_3UTR[grepl("Ago2_both|Ago2_aegypti", aegypti_Ago2_3UTR$Ptop_smallRNA),]
aegypti_6mer_Ago2_no_chim  <- aegypti_6mer_Ago2[!aegypti_6mer_Ago2$peakID %in% aegypti_perf$peakID,]
aegypti_6mer_Ago2_no_chim  <- aegypti_6mer_Ago2_no_chim [!aegypti_6mer_Ago2_no_chim $peakID %in% aegypti_chim_Ago2$peakID,]

#8mer targets
seed_table <- read.delim("/Users/kathryn/Reprocess_all_paper_datasets/Supp_Figs/all_putative_known_sRNA_seeds.txt", header = TRUE, sep ="\t", stringsAsFactors = FALSE) #913 unique seed combos
seeds_df <- read.delim("/Users/kathryn/Reprocess_all_paper_datasets/Supp_tables/Table_S6_Resource_paper_FINAL_rn.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

#link to small RNA names
seed_table <- merge(seeds_df[,c(1:2, 34)], seed_table,by.x = "sequence" , by.y = "FL", all.x= TRUE )

#select top most abundant/targeting esiRNAs
Ago2_sRNAs <- readDNAStringSet("/Users/kathryn/Reprocess_all_paper_datasets/seed_search_refseqs/FINAL/all_Ago2_known_novel_rn.fa", format = "fasta")
Ago2_sRNAs_aegypti <- Ago2_sRNAs[grepl("common-Ago2|aegypti-Ago2|aae-miR", names(Ago2_sRNAs))]

seed_table$filt <- seed_table$six_mer_target %in% Ago2_sRNAs_aegypti
Ago2_top_seeds <- subset(seed_table, seed_table$filt==TRUE)

#make vector of top Ago2 esiRNA names to search in 8mer column
vec8 <- as.character(Ago2_top_seeds$smallRNA)

row.names(aegypti_6mer_Ago2_no_chim) <- 1:nrow(aegypti_6mer_Ago2_no_chim)

matmatch8 <- vector("list", length(vec8))
for (i in 1:length(vec8)) {
  matmatch8[[i]] <- aegypti_6mer_Ago2_no_chim[grepl(vec8[i], aegypti_6mer_Ago2_no_chim$Peight_target),]
  matmatch8[[i]] <- matmatch8[[i]]$peakID
}

matmatch8 <- unique(unlist(matmatch8))
aegypti_Ago2_8 <- aegypti_6mer_Ago2_no_chim[aegypti_6mer_Ago2_no_chim$peakID %in% matmatch8,]
#have already removed perfect and chimera

##alt 8 
matmatch8alt <- vector("list", length(vec8))
for (i in 1:length(vec8)) {
  matmatch8alt[[i]] <- aegypti_6mer_Ago2_no_chim[grepl(vec8[i], aegypti_6mer_Ago2_no_chim$Peight_target_alt),]
  matmatch8alt[[i]] <- matmatch8alt[[i]]$peakID
}

matmatch8alt <- unique(unlist(matmatch8alt))
aegypti_Ago2_8alt <- aegypti_6mer_Ago2_no_chim[aegypti_6mer_Ago2_no_chim$peakID %in% matmatch8alt,]

#remove 8mers from 6mer no chim
aegypti_6mer_Ago2_no_chim_no8 <- aegypti_6mer_Ago2_no_chim[!aegypti_6mer_Ago2_no_chim$peakID %in% aegypti_Ago2_8alt$peakID,]
aegypti_6mer_Ago2_no_chim_no8 <- aegypti_6mer_Ago2_no_chim_no8[!aegypti_6mer_Ago2_no_chim_no8$peakID %in% aegypti_Ago2_8$peakID,]


#get non_3UTR 
aegypti_Ago2_non3UTR <- all_merge_aegypti_Ago2[!grepl( "3' UTR", all_merge_aegypti_Ago2$annotation),]
#is mutually exclusive with 6mers because they have to be in 3UTR, same for chimera
aegypti_Ago2_non3UTR <- aegypti_Ago2_non3UTR[!aegypti_Ago2_non3UTR$peakID %in% aegypti_perf$peakID,]
aegypti_Ago2_non3UTRtop <- aegypti_Ago2_non3UTR[grepl("Ago2_both|Ago2_aegypti", aegypti_Ago2_non3UTR$Ptop_smallRNA) ,]

tosel <- c("aegypti_Ago2_non3UTRtop", "aegypti_Ago2_8alt", "aegypti_Ago2_8", "aegypti_chim_Ago2", "aegypti_perf", "aegypti_6mer_Ago2_no_chim_no8" )


h<- do.call("list",mget(tosel))
h <- lapply(h, function(x) x[,c("geneId", "transcriptId")])
hu <- lapply(h, unique)
names(hu) <- names(h)


#for(i in seq_along(hu)) {
#  write.table(hu[i], paste0("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Ago2_lists_FINAL/", names(hu)[i], ".txt"), 
#              col.names = TRUE, row.names = FALSE, sep = "\t", quote = FALSE)
#}

```

##Now GSVA for Ago2, Figure S6B

```{r}
myAgo2List <- dir("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Ago2_lists_FINAL/",full.names = TRUE) %>% lapply(function(x)read.delim(x,header=TRUE)[,1] %>% as.vector)

#map to their geneIDs
Ago2setList <- lapply(myAgo2List,function(x) unique(annotation[annotation[,2] %in% x,1]) %>% na.omit %>% as.vector)
names(Ago2setList) <- dir("/Users/kathryn/Reprocess_all_paper_datasets/RNAseq_test/Ago2_lists_FINAL/") %>% gsub("\\.txt","",.)

removetarg <- unlist(Ago2setList)
names(removetarg) <- NULL
nottarg <- row.names(myNewCounts2)[!row.names(myNewCounts2) %in% removetarg]
Ago2setList$nontarget <- nottarg 

Ago2setList$all8 <- unique(c(Ago2setList$aegypti_Ago2_8alt, Ago2setList$aegypti_Ago2_8))

mhCountsAgo2 <- gsva(myCounts2[,colnames(myCounts2) %in% Fe],gset.idx.list = Ago2setList[!grepl("aegypti_Ago2_8alt|aegypti_Ago2_8", names(Ago2setList))],kcdf=c("Poisson"))


#visualize with heatmap
colors <- metaData2
colors <- as.data.frame(unique(colors$Group))
colors <- setNames(colors, c("Group"))
colors$color <- ifelse(grepl("Ov_O", colors$Group), paste0("lightslateblue"), paste0(colors$color))
colors$color <- ifelse(grepl("Ov_SF", colors$Group), paste0("plum"), paste0(colors$color))
mycolors <- as.character(colors$color) 
names(mycolors) <- colors$Group
mycolors <- list(Group = mycolors)
mycolors$Paired <- c("gray50", "black")
names(mycolors$Paired) <- c("Paired", "Single")

paletteLength <- 50
palAgo2 <- colorRampPalette(c("darkred", "white", "darkorange"))(paletteLength)
myBreaks <- c(seq(min(mhCountsAgo2), 0, length.out=ceiling(paletteLength/2) + 1), seq(max(mhCountsAgo2)/paletteLength, max(mhCountsAgo2), length.out=floor(paletteLength/2)))


pheatmap:::pheatmap(mhCountsAgo2,
                    annotation_col = metaData2,
                     color = palAgo2, breaks = myBreaks,
                    cluster_cols = TRUE,cluster_rows = TRUE, scale="none", fontsize_col = 6, annotation_colors = mycolors, treeheight_row = 20, fontsize = 6, cellheight = 25, cellwidth = 6)

```

##Ago2 eCDF, Figure S6E-F

```{r}
#calculate log2FC of all tissues versus ovary, bcasue see xpression in the ovary of esiRNAs 
myCountstest$gene <- row.names(myCountstest)
perf_cdf <- myCountstest[row.names(myCountstest) %in% Ago2setList$aegypti_perf,]
perf_cdf<- perf_cdf[grepl("av|mean|gene", colnames(perf_cdf))]
perf_cdf$log2perf_Ov_SF <- log2(perf_cdf$Fe_Ov_SF_av) - log2(perf_cdf$Fe_mean)
perf_cdf$log2perf_Ov_O <- log2(perf_cdf$Fe_Ov_O_av) - log2(perf_cdf$Fe_mean)
perf_cdf <- melt(perf_cdf, id.vars = c("gene"), measure.vars = c("log2perf_Ov_SF", "log2perf_Ov_O"))

six_cdf <- myCountstest[row.names(myCountstest) %in% Ago2setList$aegypti_6mer_Ago2_no_chim_no8,]
six_cdf <- six_cdf[grepl("av|mean|gene", colnames(six_cdf))]
six_cdf$log2six_Ov_SF <- log2(six_cdf$Fe_Ov_SF_av) - log2(six_cdf$Fe_mean)
six_cdf$log2six_Ov_O <- log2(six_cdf$Fe_Ov_O_av) - log2(six_cdf$Fe_mean)
six_cdf <- melt(six_cdf, id.vars = c("gene"), measure.vars = c("log2six_Ov_SF", "log2six_Ov_O"))

eight_cdf <- myCountstest[row.names(myCountstest) %in% Ago2setList$all8,]
eight_cdf <- eight_cdf[grepl("av|mean|gene", colnames(eight_cdf))]
eight_cdf$log2eight_Ov_SF <- log2(eight_cdf$Fe_Ov_SF_av) - log2(eight_cdf$Fe_mean)
eight_cdf$log2eight_Ov_O <- log2(eight_cdf$Fe_Ov_O_av) - log2(eight_cdf$Fe_mean)
eight_cdf <- melt(eight_cdf, id.vars = c("gene"), measure.vars = c("log2eight_Ov_SF", "log2eight_Ov_O"))

chim_cdf <- myCountstest[row.names(myCountstest) %in% Ago2setList$aegypti_chim_Ago2,]
chim_cdf <- chim_cdf[grepl("av|mean|gene", colnames(chim_cdf))]
chim_cdf$log2chim_Ov_SF <- log2(chim_cdf$Fe_Ov_SF_av) - log2(chim_cdf$Fe_mean)
chim_cdf$log2chim_Ov_O <- log2(chim_cdf$Fe_Ov_O_av) - log2(chim_cdf$Fe_mean)
chim_cdf <- melt(chim_cdf, id.vars = c("gene"), measure.vars = c("log2chim_Ov_SF", "log2chim_Ov_O"))

nonUTR_cdf <- myCountstest[row.names(myCountstest) %in% Ago2setList$aegypti_Ago2_non3UTRtop,]
nonUTR_cdf <- nonUTR_cdf[grepl("av|mean|gene", colnames(nonUTR_cdf))]
nonUTR_cdf$log2nonUTR_Ov_SF <- log2(nonUTR_cdf$Fe_Ov_SF_av) - log2(nonUTR_cdf$Fe_mean)
nonUTR_cdf$log2nonUTR_Ov_O <- log2(nonUTR_cdf$Fe_Ov_O_av) - log2(nonUTR_cdf$Fe_mean)
nonUTR_cdf <- melt(nonUTR_cdf, id.vars = c("gene"), measure.vars = c("log2nonUTR_Ov_SF", "log2nonUTR_Ov_O"))

Ago2_cdf <- rbind(nonUTR_cdf, six_cdf, perf_cdf, chim_cdf, eight_cdf )


ggplot(Ago2_cdf[grepl("Ov_O", Ago2_cdf$variable),], aes(value, colour = variable)) + stat_ecdf(geom = "point", size=1) + scale_colour_manual(values=c("gray","lightpink", "red4", "red2", "hotpink")) + theme_bw()

ggplot(Ago2_cdf[grepl("Ov_SF", Ago2_cdf$variable),], aes(value, colour = variable)) + stat_ecdf(geom = "point", size=1) + scale_colour_manual(values=c("gray","lightpink", "red4", "red2", "hotpink" )) + theme_bw()

ks.test(subset(Ago2_cdf$value,Ago2_cdf$variable=="log2perf_Ov_SF"), subset(Ago2_cdf$value, Ago2_cdf$variable=="log2nonUTR_Ov_SF"),alternative = c("greater")) #0.003818
ks.test(subset(Ago2_cdf$value,Ago2_cdf$variable=="log2six_Ov_SF"), subset(Ago2_cdf$value, Ago2_cdf$variable=="log2nonUTR_Ov_SF"),alternative = c("greater")) #0.2691
ks.test(subset(Ago2_cdf$value,Ago2_cdf$variable=="log2eight_Ov_SF"), subset(Ago2_cdf$value, Ago2_cdf$variable=="log2nonUTR_Ov_SF"),alternative = c("g")) #0.04076
ks.test(subset(Ago2_cdf$value,Ago2_cdf$variable=="log2chim_Ov_SF"), subset(Ago2_cdf$value, Ago2_cdf$variable=="log2nonUTR_Ov_SF"),alternative = c("greater")) #0.005088

ks.test(subset(Ago2_cdf$value,Ago2_cdf$variable=="log2perf_Ov_O"), subset(Ago2_cdf$value, Ago2_cdf$variable=="log2nonUTR_Ov_O"),alternative = c("greater")) #0.007697
ks.test(subset(Ago2_cdf$value,Ago2_cdf$variable=="log2six_Ov_O"), subset(Ago2_cdf$value, Ago2_cdf$variable=="log2nonUTR_Ov_O"),alternative = c("greater")) # 0.2971
ks.test(subset(Ago2_cdf$value,Ago2_cdf$variable=="log2eight_Ov_O"), subset(Ago2_cdf$value, Ago2_cdf$variable=="log2nonUTR_Ov_O"),alternative = c("greater")) # 0.1324
ks.test(subset(Ago2_cdf$value,Ago2_cdf$variable=="log2chim_Ov_O"), subset(Ago2_cdf$value, Ago2_cdf$variable=="log2nonUTR_Ov_O"),alternative = c("greater")) #0.001285

```
